// miniads/dist/cjs/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n["default"] = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespace(React);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

/******************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
var __assign = function () {
  __assign = Object.assign || function __assign(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
      s = arguments[i];
      for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
    }
    return t;
  };
  return __assign.apply(this, arguments);
};
function __awaiter(thisArg, _arguments, P, generator) {
  function adopt(value) {
    return value instanceof P ? value : new P(function (resolve) {
      resolve(value);
    });
  }
  return new (P || (P = Promise))(function (resolve, reject) {
    function fulfilled(value) {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    }
    function rejected(value) {
      try {
        step(generator["throw"](value));
      } catch (e) {
        reject(e);
      }
    }
    function step(result) {
      result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected);
    }
    step((generator = generator.apply(thisArg, _arguments || [])).next());
  });
}
function __generator(thisArg, body) {
  var _ = {
      label: 0,
      sent: function () {
        if (t[0] & 1) throw t[1];
        return t[1];
      },
      trys: [],
      ops: []
    },
    f,
    y,
    t,
    g;
  return g = {
    next: verb(0),
    "throw": verb(1),
    "return": verb(2)
  }, typeof Symbol === "function" && (g[Symbol.iterator] = function () {
    return this;
  }), g;
  function verb(n) {
    return function (v) {
      return step([n, v]);
    };
  }
  function step(op) {
    if (f) throw new TypeError("Generator is already executing.");
    while (g && (g = 0, op[0] && (_ = 0)), _) try {
      if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
      if (y = 0, t) op = [op[0] & 2, t.value];
      switch (op[0]) {
        case 0:
        case 1:
          t = op;
          break;
        case 4:
          _.label++;
          return {
            value: op[1],
            done: false
          };
        case 5:
          _.label++;
          y = op[1];
          op = [0];
          continue;
        case 7:
          op = _.ops.pop();
          _.trys.pop();
          continue;
        default:
          if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
            _ = 0;
            continue;
          }
          if (op[0] === 3 && (!t || op[1] > t[0] && op[1] < t[3])) {
            _.label = op[1];
            break;
          }
          if (op[0] === 6 && _.label < t[1]) {
            _.label = t[1];
            t = op;
            break;
          }
          if (t && _.label < t[2]) {
            _.label = t[2];
            _.ops.push(op);
            break;
          }
          if (t[2]) _.ops.pop();
          _.trys.pop();
          continue;
      }
      op = body.call(thisArg, _);
    } catch (e) {
      op = [6, e];
      y = 0;
    } finally {
      f = t = 0;
    }
    if (op[0] & 5) throw op[1];
    return {
      value: op[0] ? op[1] : void 0,
      done: true
    };
  }
}
function __makeTemplateObject(cooked, raw) {
  if (Object.defineProperty) {
    Object.defineProperty(cooked, "raw", {
      value: raw
    });
  } else {
    cooked.raw = raw;
  }
  return cooked;
}
typeof SuppressedError === "function" ? SuppressedError : function (error, suppressed, message) {
  var e = new Error(message);
  return e.name = "SuppressedError", e.error = error, e.suppressed = suppressed, e;
};

var SvgArrow = function (props) { return (React__namespace.createElement("svg", __assign({ fill: "none", height: 40, viewBox: "0 0 40 40", width: 40, xmlns: "http://www.w3.org/2000/svg" }, props),
    React__namespace.createElement("path", { d: "M39.7354 10.8671C39.7354 10.4737 39.5852 10.0803 39.2853 9.78038C38.6856 9.18018 37.7122 9.18018 37.1124 9.78038L19.9776 26.9148L2.84315 9.78038C2.24295 9.18018 1.26999 9.18018 0.66979 9.78038C0.0700308 10.3806 0.0700308 11.3535 0.66979 11.9537L18.8909 30.1748C19.4906 30.775 20.464 30.775 21.0638 30.1748L39.2849 11.9537C39.5852 11.6539 39.7354 11.2605 39.7354 10.8671Z", fill: "black" }))); };
React.memo(SvgArrow);

var SvgUnMute = function (props) { return (React__namespace.createElement("svg", __assign({ xmlns: "http://www.w3.org/2000/svg", width: 16, height: 16, viewBox: "0 0 16 16" }, props),
    React__namespace.createElement("path", { fillRule: "evenodd", d: "M12 8.02c0 1.09-.45 2.09-1.17 2.83l-.67-.67c.55-.56.89-1.31.89-2.16 0-.85-.34-1.61-.89-2.16l.67-.67A3.99 3.99 0 0 1 12 8.02zM7.72 2.28L4 6H2c-.55 0-1 .45-1 1v2c0 .55.45 1 1 1h2l3.72 3.72c.47.47 1.28.14 1.28-.53V2.81c0-.67-.81-1-1.28-.53zm5.94.08l-.67.67a6.996 6.996 0 0 1 2.06 4.98c0 1.94-.78 3.7-2.06 4.98l.67.67A7.973 7.973 0 0 0 16 8c0-2.22-.89-4.22-2.34-5.66v.02zm-1.41 1.41l-.69.67a5.05 5.05 0 0 1 1.48 3.58c0 1.39-.56 2.66-1.48 3.56l.69.67A5.97 5.97 0 0 0 14 8.02c0-1.65-.67-3.16-1.75-4.25z" }))); };
var Memo$3 = React.memo(SvgUnMute);

var SvgMute = function (props) { return (React__namespace.createElement("svg", __assign({ xmlns: "http://www.w3.org/2000/svg", width: 16, height: 16, viewBox: "0 0 16 16" }, props),
    React__namespace.createElement("path", { fillRule: "evenodd", d: "M8 2.81v10.38c0 .67-.81 1-1.28.53L3 10H1c-.55 0-1-.45-1-1V7c0-.55.45-1 1-1h2l3.72-3.72C7.19 1.81 8 2.14 8 2.81zm7.53 3.22l-1.06-1.06-1.97 1.97-1.97-1.97-1.06 1.06L11.44 8 9.47 9.97l1.06 1.06 1.97-1.97 1.97 1.97 1.06-1.06L13.56 8l1.97-1.97z" }))); };
var Memo$2 = React.memo(SvgMute);

var SvgCross = function (props) { return (React__namespace.createElement("svg", { width: 12, height: 12, viewBox: "0 0 12 12", fill: "none", xmlns: "http://www.w3.org/2000/svg" },
    React__namespace.createElement("path", { fillRule: "evenodd", d: "M0.637695 11.3773C0.515951 11.2611 0.43571 11.1228 0.396973 10.9623C0.358236 10.7963 0.358236 10.633 0.396973 10.4725C0.441243 10.3121 0.521484 10.1765 0.637695 10.0658L4.68018 6.01501L0.637695 1.97253C0.521484 1.86186 0.441243 1.72628 0.396973 1.5658C0.358236 1.40531 0.358236 1.24483 0.396973 1.08435C0.43571 0.918335 0.515951 0.777222 0.637695 0.661011C0.75944 0.539266 0.900553 0.459025 1.06104 0.420288C1.22152 0.381551 1.382 0.384318 1.54248 0.428589C1.70296 0.467326 1.84131 0.542033 1.95752 0.65271L6 4.70349L10.0425 0.65271C10.1587 0.536499 10.297 0.459025 10.4575 0.420288C10.618 0.381551 10.7757 0.384318 10.9307 0.428589C11.0911 0.467326 11.2323 0.5448 11.354 0.661011C11.4757 0.777222 11.556 0.918335 11.5947 1.08435C11.6335 1.24483 11.6335 1.40531 11.5947 1.5658C11.556 1.72074 11.4757 1.85909 11.354 1.98083L7.31982 6.01501L11.354 10.0658C11.4757 10.182 11.556 10.3204 11.5947 10.4808C11.6335 10.6358 11.6335 10.7963 11.5947 10.9623C11.556 11.1228 11.4757 11.2611 11.354 11.3773C11.2323 11.4935 11.0911 11.5738 10.9307 11.618C10.7757 11.6568 10.618 11.6568 10.4575 11.618C10.297 11.5793 10.1587 11.5018 10.0425 11.3856L6 7.33484L1.95752 11.3856C1.84131 11.5018 1.70296 11.5793 1.54248 11.618C1.382 11.6568 1.22152 11.6568 1.06104 11.618C0.900553 11.5793 0.75944 11.4991 0.637695 11.3773Z", fill: "#8E8E93" }))); };
var Memo$1 = React.memo(SvgCross);

var SvgCrossEditable = function (props) { return (React__namespace.createElement("svg", __assign({ viewBox: "0 0 12 12", width: 12, height: 12, xmlns: "http://www.w3.org/2000/svg" }, props),
    React__namespace.createElement("path", { fillRule: "evenodd", d: "M0.637695 11.3773C0.515951 11.2611 0.43571 11.1228 0.396973 10.9623C0.358236 10.7963 0.358236 10.633 0.396973 10.4725C0.441243 10.3121 0.521484 10.1765 0.637695 10.0658L4.68018 6.01501L0.637695 1.97253C0.521484 1.86186 0.441243 1.72628 0.396973 1.5658C0.358236 1.40531 0.358236 1.24483 0.396973 1.08435C0.43571 0.918335 0.515951 0.777222 0.637695 0.661011C0.75944 0.539266 0.900553 0.459025 1.06104 0.420288C1.22152 0.381551 1.382 0.384318 1.54248 0.428589C1.70296 0.467326 1.84131 0.542033 1.95752 0.65271L6 4.70349L10.0425 0.65271C10.1587 0.536499 10.297 0.459025 10.4575 0.420288C10.618 0.381551 10.7757 0.384318 10.9307 0.428589C11.0911 0.467326 11.2323 0.5448 11.354 0.661011C11.4757 0.777222 11.556 0.918335 11.5947 1.08435C11.6335 1.24483 11.6335 1.40531 11.5947 1.5658C11.556 1.72074 11.4757 1.85909 11.354 1.98083L7.31982 6.01501L11.354 10.0658C11.4757 10.182 11.556 10.3204 11.5947 10.4808C11.6335 10.6358 11.6335 10.7963 11.5947 10.9623C11.556 11.1228 11.4757 11.2611 11.354 11.3773C11.2323 11.4935 11.0911 11.5738 10.9307 11.618C10.7757 11.6568 10.618 11.6568 10.4575 11.618C10.297 11.5793 10.1587 11.5018 10.0425 11.3856L6 7.33484L1.95752 11.3856C1.84131 11.5018 1.70296 11.5793 1.54248 11.618C1.382 11.6568 1.22152 11.6568 1.06104 11.618C0.900553 11.5793 0.75944 11.4991 0.637695 11.3773Z" }))); };
var Memo = React.memo(SvgCrossEditable);

var reactIs$1 = {exports: {}};

var reactIs_production_min = {};

/** @license React v16.13.1
 * react-is.production.min.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var b$1 = "function" === typeof Symbol && Symbol.for,
  c = b$1 ? Symbol.for("react.element") : 60103,
  d = b$1 ? Symbol.for("react.portal") : 60106,
  e = b$1 ? Symbol.for("react.fragment") : 60107,
  f = b$1 ? Symbol.for("react.strict_mode") : 60108,
  g$1 = b$1 ? Symbol.for("react.profiler") : 60114,
  h = b$1 ? Symbol.for("react.provider") : 60109,
  k$1 = b$1 ? Symbol.for("react.context") : 60110,
  l = b$1 ? Symbol.for("react.async_mode") : 60111,
  m = b$1 ? Symbol.for("react.concurrent_mode") : 60111,
  n = b$1 ? Symbol.for("react.forward_ref") : 60112,
  p = b$1 ? Symbol.for("react.suspense") : 60113,
  q$1 = b$1 ? Symbol.for("react.suspense_list") : 60120,
  r = b$1 ? Symbol.for("react.memo") : 60115,
  t = b$1 ? Symbol.for("react.lazy") : 60116,
  v$1 = b$1 ? Symbol.for("react.block") : 60121,
  w$1 = b$1 ? Symbol.for("react.fundamental") : 60117,
  x$1 = b$1 ? Symbol.for("react.responder") : 60118,
  y$1 = b$1 ? Symbol.for("react.scope") : 60119;
function z$1(a) {
  if ("object" === typeof a && null !== a) {
    var u = a.$$typeof;
    switch (u) {
      case c:
        switch (a = a.type, a) {
          case l:
          case m:
          case e:
          case g$1:
          case f:
          case p:
            return a;
          default:
            switch (a = a && a.$$typeof, a) {
              case k$1:
              case n:
              case t:
              case r:
              case h:
                return a;
              default:
                return u;
            }
        }
      case d:
        return u;
    }
  }
}
function A(a) {
  return z$1(a) === m;
}
reactIs_production_min.AsyncMode = l;
reactIs_production_min.ConcurrentMode = m;
reactIs_production_min.ContextConsumer = k$1;
reactIs_production_min.ContextProvider = h;
reactIs_production_min.Element = c;
reactIs_production_min.ForwardRef = n;
reactIs_production_min.Fragment = e;
reactIs_production_min.Lazy = t;
reactIs_production_min.Memo = r;
reactIs_production_min.Portal = d;
reactIs_production_min.Profiler = g$1;
reactIs_production_min.StrictMode = f;
reactIs_production_min.Suspense = p;
reactIs_production_min.isAsyncMode = function (a) {
  return A(a) || z$1(a) === l;
};
reactIs_production_min.isConcurrentMode = A;
reactIs_production_min.isContextConsumer = function (a) {
  return z$1(a) === k$1;
};
reactIs_production_min.isContextProvider = function (a) {
  return z$1(a) === h;
};
reactIs_production_min.isElement = function (a) {
  return "object" === typeof a && null !== a && a.$$typeof === c;
};
reactIs_production_min.isForwardRef = function (a) {
  return z$1(a) === n;
};
reactIs_production_min.isFragment = function (a) {
  return z$1(a) === e;
};
reactIs_production_min.isLazy = function (a) {
  return z$1(a) === t;
};
reactIs_production_min.isMemo = function (a) {
  return z$1(a) === r;
};
reactIs_production_min.isPortal = function (a) {
  return z$1(a) === d;
};
reactIs_production_min.isProfiler = function (a) {
  return z$1(a) === g$1;
};
reactIs_production_min.isStrictMode = function (a) {
  return z$1(a) === f;
};
reactIs_production_min.isSuspense = function (a) {
  return z$1(a) === p;
};
reactIs_production_min.isValidElementType = function (a) {
  return "string" === typeof a || "function" === typeof a || a === e || a === m || a === g$1 || a === f || a === p || a === q$1 || "object" === typeof a && null !== a && (a.$$typeof === t || a.$$typeof === r || a.$$typeof === h || a.$$typeof === k$1 || a.$$typeof === n || a.$$typeof === w$1 || a.$$typeof === x$1 || a.$$typeof === y$1 || a.$$typeof === v$1);
};
reactIs_production_min.typeOf = z$1;

var reactIs_development = {};

/** @license React v16.13.1
 * react-is.development.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
if (process.env.NODE_ENV !== "production") {
  (function () {

    // The Symbol used to tag the ReactElement-like types. If there is no native Symbol
    // nor polyfill, then a plain number is used for performance.
    var hasSymbol = typeof Symbol === 'function' && Symbol.for;
    var REACT_ELEMENT_TYPE = hasSymbol ? Symbol.for('react.element') : 0xeac7;
    var REACT_PORTAL_TYPE = hasSymbol ? Symbol.for('react.portal') : 0xeaca;
    var REACT_FRAGMENT_TYPE = hasSymbol ? Symbol.for('react.fragment') : 0xeacb;
    var REACT_STRICT_MODE_TYPE = hasSymbol ? Symbol.for('react.strict_mode') : 0xeacc;
    var REACT_PROFILER_TYPE = hasSymbol ? Symbol.for('react.profiler') : 0xead2;
    var REACT_PROVIDER_TYPE = hasSymbol ? Symbol.for('react.provider') : 0xeacd;
    var REACT_CONTEXT_TYPE = hasSymbol ? Symbol.for('react.context') : 0xeace; // TODO: We don't use AsyncMode or ConcurrentMode anymore. They were temporary
    // (unstable) APIs that have been removed. Can we remove the symbols?

    var REACT_ASYNC_MODE_TYPE = hasSymbol ? Symbol.for('react.async_mode') : 0xeacf;
    var REACT_CONCURRENT_MODE_TYPE = hasSymbol ? Symbol.for('react.concurrent_mode') : 0xeacf;
    var REACT_FORWARD_REF_TYPE = hasSymbol ? Symbol.for('react.forward_ref') : 0xead0;
    var REACT_SUSPENSE_TYPE = hasSymbol ? Symbol.for('react.suspense') : 0xead1;
    var REACT_SUSPENSE_LIST_TYPE = hasSymbol ? Symbol.for('react.suspense_list') : 0xead8;
    var REACT_MEMO_TYPE = hasSymbol ? Symbol.for('react.memo') : 0xead3;
    var REACT_LAZY_TYPE = hasSymbol ? Symbol.for('react.lazy') : 0xead4;
    var REACT_BLOCK_TYPE = hasSymbol ? Symbol.for('react.block') : 0xead9;
    var REACT_FUNDAMENTAL_TYPE = hasSymbol ? Symbol.for('react.fundamental') : 0xead5;
    var REACT_RESPONDER_TYPE = hasSymbol ? Symbol.for('react.responder') : 0xead6;
    var REACT_SCOPE_TYPE = hasSymbol ? Symbol.for('react.scope') : 0xead7;
    function isValidElementType(type) {
      return typeof type === 'string' || typeof type === 'function' ||
      // Note: its typeof might be other than 'symbol' or 'number' if it's a polyfill.
      type === REACT_FRAGMENT_TYPE || type === REACT_CONCURRENT_MODE_TYPE || type === REACT_PROFILER_TYPE || type === REACT_STRICT_MODE_TYPE || type === REACT_SUSPENSE_TYPE || type === REACT_SUSPENSE_LIST_TYPE || typeof type === 'object' && type !== null && (type.$$typeof === REACT_LAZY_TYPE || type.$$typeof === REACT_MEMO_TYPE || type.$$typeof === REACT_PROVIDER_TYPE || type.$$typeof === REACT_CONTEXT_TYPE || type.$$typeof === REACT_FORWARD_REF_TYPE || type.$$typeof === REACT_FUNDAMENTAL_TYPE || type.$$typeof === REACT_RESPONDER_TYPE || type.$$typeof === REACT_SCOPE_TYPE || type.$$typeof === REACT_BLOCK_TYPE);
    }
    function typeOf(object) {
      if (typeof object === 'object' && object !== null) {
        var $$typeof = object.$$typeof;
        switch ($$typeof) {
          case REACT_ELEMENT_TYPE:
            var type = object.type;
            switch (type) {
              case REACT_ASYNC_MODE_TYPE:
              case REACT_CONCURRENT_MODE_TYPE:
              case REACT_FRAGMENT_TYPE:
              case REACT_PROFILER_TYPE:
              case REACT_STRICT_MODE_TYPE:
              case REACT_SUSPENSE_TYPE:
                return type;
              default:
                var $$typeofType = type && type.$$typeof;
                switch ($$typeofType) {
                  case REACT_CONTEXT_TYPE:
                  case REACT_FORWARD_REF_TYPE:
                  case REACT_LAZY_TYPE:
                  case REACT_MEMO_TYPE:
                  case REACT_PROVIDER_TYPE:
                    return $$typeofType;
                  default:
                    return $$typeof;
                }
            }
          case REACT_PORTAL_TYPE:
            return $$typeof;
        }
      }
      return undefined;
    } // AsyncMode is deprecated along with isAsyncMode

    var AsyncMode = REACT_ASYNC_MODE_TYPE;
    var ConcurrentMode = REACT_CONCURRENT_MODE_TYPE;
    var ContextConsumer = REACT_CONTEXT_TYPE;
    var ContextProvider = REACT_PROVIDER_TYPE;
    var Element = REACT_ELEMENT_TYPE;
    var ForwardRef = REACT_FORWARD_REF_TYPE;
    var Fragment = REACT_FRAGMENT_TYPE;
    var Lazy = REACT_LAZY_TYPE;
    var Memo = REACT_MEMO_TYPE;
    var Portal = REACT_PORTAL_TYPE;
    var Profiler = REACT_PROFILER_TYPE;
    var StrictMode = REACT_STRICT_MODE_TYPE;
    var Suspense = REACT_SUSPENSE_TYPE;
    var hasWarnedAboutDeprecatedIsAsyncMode = false; // AsyncMode should be deprecated

    function isAsyncMode(object) {
      {
        if (!hasWarnedAboutDeprecatedIsAsyncMode) {
          hasWarnedAboutDeprecatedIsAsyncMode = true; // Using console['warn'] to evade Babel and ESLint

          console['warn']('The ReactIs.isAsyncMode() alias has been deprecated, ' + 'and will be removed in React 17+. Update your code to use ' + 'ReactIs.isConcurrentMode() instead. It has the exact same API.');
        }
      }
      return isConcurrentMode(object) || typeOf(object) === REACT_ASYNC_MODE_TYPE;
    }
    function isConcurrentMode(object) {
      return typeOf(object) === REACT_CONCURRENT_MODE_TYPE;
    }
    function isContextConsumer(object) {
      return typeOf(object) === REACT_CONTEXT_TYPE;
    }
    function isContextProvider(object) {
      return typeOf(object) === REACT_PROVIDER_TYPE;
    }
    function isElement(object) {
      return typeof object === 'object' && object !== null && object.$$typeof === REACT_ELEMENT_TYPE;
    }
    function isForwardRef(object) {
      return typeOf(object) === REACT_FORWARD_REF_TYPE;
    }
    function isFragment(object) {
      return typeOf(object) === REACT_FRAGMENT_TYPE;
    }
    function isLazy(object) {
      return typeOf(object) === REACT_LAZY_TYPE;
    }
    function isMemo(object) {
      return typeOf(object) === REACT_MEMO_TYPE;
    }
    function isPortal(object) {
      return typeOf(object) === REACT_PORTAL_TYPE;
    }
    function isProfiler(object) {
      return typeOf(object) === REACT_PROFILER_TYPE;
    }
    function isStrictMode(object) {
      return typeOf(object) === REACT_STRICT_MODE_TYPE;
    }
    function isSuspense(object) {
      return typeOf(object) === REACT_SUSPENSE_TYPE;
    }
    reactIs_development.AsyncMode = AsyncMode;
    reactIs_development.ConcurrentMode = ConcurrentMode;
    reactIs_development.ContextConsumer = ContextConsumer;
    reactIs_development.ContextProvider = ContextProvider;
    reactIs_development.Element = Element;
    reactIs_development.ForwardRef = ForwardRef;
    reactIs_development.Fragment = Fragment;
    reactIs_development.Lazy = Lazy;
    reactIs_development.Memo = Memo;
    reactIs_development.Portal = Portal;
    reactIs_development.Profiler = Profiler;
    reactIs_development.StrictMode = StrictMode;
    reactIs_development.Suspense = Suspense;
    reactIs_development.isAsyncMode = isAsyncMode;
    reactIs_development.isConcurrentMode = isConcurrentMode;
    reactIs_development.isContextConsumer = isContextConsumer;
    reactIs_development.isContextProvider = isContextProvider;
    reactIs_development.isElement = isElement;
    reactIs_development.isForwardRef = isForwardRef;
    reactIs_development.isFragment = isFragment;
    reactIs_development.isLazy = isLazy;
    reactIs_development.isMemo = isMemo;
    reactIs_development.isPortal = isPortal;
    reactIs_development.isProfiler = isProfiler;
    reactIs_development.isStrictMode = isStrictMode;
    reactIs_development.isSuspense = isSuspense;
    reactIs_development.isValidElementType = isValidElementType;
    reactIs_development.typeOf = typeOf;
  })();
}

if (process.env.NODE_ENV === 'production') {
  reactIs$1.exports = reactIs_production_min;
} else {
  reactIs$1.exports = reactIs_development;
}

function stylis_min(W) {
  function M(d, c, e, h, a) {
    for (var m = 0, b = 0, v = 0, n = 0, q, g, x = 0, K = 0, k, u = k = q = 0, l = 0, r = 0, I = 0, t = 0, B = e.length, J = B - 1, y, f = '', p = '', F = '', G = '', C; l < B;) {
      g = e.charCodeAt(l);
      l === J && 0 !== b + n + v + m && (0 !== b && (g = 47 === b ? 10 : 47), n = v = m = 0, B++, J++);
      if (0 === b + n + v + m) {
        if (l === J && (0 < r && (f = f.replace(N, '')), 0 < f.trim().length)) {
          switch (g) {
            case 32:
            case 9:
            case 59:
            case 13:
            case 10:
              break;
            default:
              f += e.charAt(l);
          }
          g = 59;
        }
        switch (g) {
          case 123:
            f = f.trim();
            q = f.charCodeAt(0);
            k = 1;
            for (t = ++l; l < B;) {
              switch (g = e.charCodeAt(l)) {
                case 123:
                  k++;
                  break;
                case 125:
                  k--;
                  break;
                case 47:
                  switch (g = e.charCodeAt(l + 1)) {
                    case 42:
                    case 47:
                      a: {
                        for (u = l + 1; u < J; ++u) {
                          switch (e.charCodeAt(u)) {
                            case 47:
                              if (42 === g && 42 === e.charCodeAt(u - 1) && l + 2 !== u) {
                                l = u + 1;
                                break a;
                              }
                              break;
                            case 10:
                              if (47 === g) {
                                l = u + 1;
                                break a;
                              }
                          }
                        }
                        l = u;
                      }
                  }
                  break;
                case 91:
                  g++;
                case 40:
                  g++;
                case 34:
                case 39:
                  for (; l++ < J && e.charCodeAt(l) !== g;) {}
              }
              if (0 === k) break;
              l++;
            }
            k = e.substring(t, l);
            0 === q && (q = (f = f.replace(ca, '').trim()).charCodeAt(0));
            switch (q) {
              case 64:
                0 < r && (f = f.replace(N, ''));
                g = f.charCodeAt(1);
                switch (g) {
                  case 100:
                  case 109:
                  case 115:
                  case 45:
                    r = c;
                    break;
                  default:
                    r = O;
                }
                k = M(c, r, k, g, a + 1);
                t = k.length;
                0 < A && (r = X(O, f, I), C = H(3, k, r, c, D, z, t, g, a, h), f = r.join(''), void 0 !== C && 0 === (t = (k = C.trim()).length) && (g = 0, k = ''));
                if (0 < t) switch (g) {
                  case 115:
                    f = f.replace(da, ea);
                  case 100:
                  case 109:
                  case 45:
                    k = f + '{' + k + '}';
                    break;
                  case 107:
                    f = f.replace(fa, '$1 $2');
                    k = f + '{' + k + '}';
                    k = 1 === w || 2 === w && L('@' + k, 3) ? '@-webkit-' + k + '@' + k : '@' + k;
                    break;
                  default:
                    k = f + k, 112 === h && (k = (p += k, ''));
                } else k = '';
                break;
              default:
                k = M(c, X(c, f, I), k, h, a + 1);
            }
            F += k;
            k = I = r = u = q = 0;
            f = '';
            g = e.charCodeAt(++l);
            break;
          case 125:
          case 59:
            f = (0 < r ? f.replace(N, '') : f).trim();
            if (1 < (t = f.length)) switch (0 === u && (q = f.charCodeAt(0), 45 === q || 96 < q && 123 > q) && (t = (f = f.replace(' ', ':')).length), 0 < A && void 0 !== (C = H(1, f, c, d, D, z, p.length, h, a, h)) && 0 === (t = (f = C.trim()).length) && (f = '\x00\x00'), q = f.charCodeAt(0), g = f.charCodeAt(1), q) {
              case 0:
                break;
              case 64:
                if (105 === g || 99 === g) {
                  G += f + e.charAt(l);
                  break;
                }
              default:
                58 !== f.charCodeAt(t - 1) && (p += P(f, q, g, f.charCodeAt(2)));
            }
            I = r = u = q = 0;
            f = '';
            g = e.charCodeAt(++l);
        }
      }
      switch (g) {
        case 13:
        case 10:
          47 === b ? b = 0 : 0 === 1 + q && 107 !== h && 0 < f.length && (r = 1, f += '\x00');
          0 < A * Y && H(0, f, c, d, D, z, p.length, h, a, h);
          z = 1;
          D++;
          break;
        case 59:
        case 125:
          if (0 === b + n + v + m) {
            z++;
            break;
          }
        default:
          z++;
          y = e.charAt(l);
          switch (g) {
            case 9:
            case 32:
              if (0 === n + m + b) switch (x) {
                case 44:
                case 58:
                case 9:
                case 32:
                  y = '';
                  break;
                default:
                  32 !== g && (y = ' ');
              }
              break;
            case 0:
              y = '\\0';
              break;
            case 12:
              y = '\\f';
              break;
            case 11:
              y = '\\v';
              break;
            case 38:
              0 === n + b + m && (r = I = 1, y = '\f' + y);
              break;
            case 108:
              if (0 === n + b + m + E && 0 < u) switch (l - u) {
                case 2:
                  112 === x && 58 === e.charCodeAt(l - 3) && (E = x);
                case 8:
                  111 === K && (E = K);
              }
              break;
            case 58:
              0 === n + b + m && (u = l);
              break;
            case 44:
              0 === b + v + n + m && (r = 1, y += '\r');
              break;
            case 34:
            case 39:
              0 === b && (n = n === g ? 0 : 0 === n ? g : n);
              break;
            case 91:
              0 === n + b + v && m++;
              break;
            case 93:
              0 === n + b + v && m--;
              break;
            case 41:
              0 === n + b + m && v--;
              break;
            case 40:
              if (0 === n + b + m) {
                if (0 === q) switch (2 * x + 3 * K) {
                  case 533:
                    break;
                  default:
                    q = 1;
                }
                v++;
              }
              break;
            case 64:
              0 === b + v + n + m + u + k && (k = 1);
              break;
            case 42:
            case 47:
              if (!(0 < n + m + v)) switch (b) {
                case 0:
                  switch (2 * g + 3 * e.charCodeAt(l + 1)) {
                    case 235:
                      b = 47;
                      break;
                    case 220:
                      t = l, b = 42;
                  }
                  break;
                case 42:
                  47 === g && 42 === x && t + 2 !== l && (33 === e.charCodeAt(t + 2) && (p += e.substring(t, l + 1)), y = '', b = 0);
              }
          }
          0 === b && (f += y);
      }
      K = x;
      x = g;
      l++;
    }
    t = p.length;
    if (0 < t) {
      r = c;
      if (0 < A && (C = H(2, p, r, d, D, z, t, h, a, h), void 0 !== C && 0 === (p = C).length)) return G + p + F;
      p = r.join(',') + '{' + p + '}';
      if (0 !== w * E) {
        2 !== w || L(p, 2) || (E = 0);
        switch (E) {
          case 111:
            p = p.replace(ha, ':-moz-$1') + p;
            break;
          case 112:
            p = p.replace(Q, '::-webkit-input-$1') + p.replace(Q, '::-moz-$1') + p.replace(Q, ':-ms-input-$1') + p;
        }
        E = 0;
      }
    }
    return G + p + F;
  }
  function X(d, c, e) {
    var h = c.trim().split(ia);
    c = h;
    var a = h.length,
      m = d.length;
    switch (m) {
      case 0:
      case 1:
        var b = 0;
        for (d = 0 === m ? '' : d[0] + ' '; b < a; ++b) {
          c[b] = Z(d, c[b], e).trim();
        }
        break;
      default:
        var v = b = 0;
        for (c = []; b < a; ++b) {
          for (var n = 0; n < m; ++n) {
            c[v++] = Z(d[n] + ' ', h[b], e).trim();
          }
        }
    }
    return c;
  }
  function Z(d, c, e) {
    var h = c.charCodeAt(0);
    33 > h && (h = (c = c.trim()).charCodeAt(0));
    switch (h) {
      case 38:
        return c.replace(F, '$1' + d.trim());
      case 58:
        return d.trim() + c.replace(F, '$1' + d.trim());
      default:
        if (0 < 1 * e && 0 < c.indexOf('\f')) return c.replace(F, (58 === d.charCodeAt(0) ? '' : '$1') + d.trim());
    }
    return d + c;
  }
  function P(d, c, e, h) {
    var a = d + ';',
      m = 2 * c + 3 * e + 4 * h;
    if (944 === m) {
      d = a.indexOf(':', 9) + 1;
      var b = a.substring(d, a.length - 1).trim();
      b = a.substring(0, d).trim() + b + ';';
      return 1 === w || 2 === w && L(b, 1) ? '-webkit-' + b + b : b;
    }
    if (0 === w || 2 === w && !L(a, 1)) return a;
    switch (m) {
      case 1015:
        return 97 === a.charCodeAt(10) ? '-webkit-' + a + a : a;
      case 951:
        return 116 === a.charCodeAt(3) ? '-webkit-' + a + a : a;
      case 963:
        return 110 === a.charCodeAt(5) ? '-webkit-' + a + a : a;
      case 1009:
        if (100 !== a.charCodeAt(4)) break;
      case 969:
      case 942:
        return '-webkit-' + a + a;
      case 978:
        return '-webkit-' + a + '-moz-' + a + a;
      case 1019:
      case 983:
        return '-webkit-' + a + '-moz-' + a + '-ms-' + a + a;
      case 883:
        if (45 === a.charCodeAt(8)) return '-webkit-' + a + a;
        if (0 < a.indexOf('image-set(', 11)) return a.replace(ja, '$1-webkit-$2') + a;
        break;
      case 932:
        if (45 === a.charCodeAt(4)) switch (a.charCodeAt(5)) {
          case 103:
            return '-webkit-box-' + a.replace('-grow', '') + '-webkit-' + a + '-ms-' + a.replace('grow', 'positive') + a;
          case 115:
            return '-webkit-' + a + '-ms-' + a.replace('shrink', 'negative') + a;
          case 98:
            return '-webkit-' + a + '-ms-' + a.replace('basis', 'preferred-size') + a;
        }
        return '-webkit-' + a + '-ms-' + a + a;
      case 964:
        return '-webkit-' + a + '-ms-flex-' + a + a;
      case 1023:
        if (99 !== a.charCodeAt(8)) break;
        b = a.substring(a.indexOf(':', 15)).replace('flex-', '').replace('space-between', 'justify');
        return '-webkit-box-pack' + b + '-webkit-' + a + '-ms-flex-pack' + b + a;
      case 1005:
        return ka.test(a) ? a.replace(aa, ':-webkit-') + a.replace(aa, ':-moz-') + a : a;
      case 1e3:
        b = a.substring(13).trim();
        c = b.indexOf('-') + 1;
        switch (b.charCodeAt(0) + b.charCodeAt(c)) {
          case 226:
            b = a.replace(G, 'tb');
            break;
          case 232:
            b = a.replace(G, 'tb-rl');
            break;
          case 220:
            b = a.replace(G, 'lr');
            break;
          default:
            return a;
        }
        return '-webkit-' + a + '-ms-' + b + a;
      case 1017:
        if (-1 === a.indexOf('sticky', 9)) break;
      case 975:
        c = (a = d).length - 10;
        b = (33 === a.charCodeAt(c) ? a.substring(0, c) : a).substring(d.indexOf(':', 7) + 1).trim();
        switch (m = b.charCodeAt(0) + (b.charCodeAt(7) | 0)) {
          case 203:
            if (111 > b.charCodeAt(8)) break;
          case 115:
            a = a.replace(b, '-webkit-' + b) + ';' + a;
            break;
          case 207:
          case 102:
            a = a.replace(b, '-webkit-' + (102 < m ? 'inline-' : '') + 'box') + ';' + a.replace(b, '-webkit-' + b) + ';' + a.replace(b, '-ms-' + b + 'box') + ';' + a;
        }
        return a + ';';
      case 938:
        if (45 === a.charCodeAt(5)) switch (a.charCodeAt(6)) {
          case 105:
            return b = a.replace('-items', ''), '-webkit-' + a + '-webkit-box-' + b + '-ms-flex-' + b + a;
          case 115:
            return '-webkit-' + a + '-ms-flex-item-' + a.replace(ba, '') + a;
          default:
            return '-webkit-' + a + '-ms-flex-line-pack' + a.replace('align-content', '').replace(ba, '') + a;
        }
        break;
      case 973:
      case 989:
        if (45 !== a.charCodeAt(3) || 122 === a.charCodeAt(4)) break;
      case 931:
      case 953:
        if (!0 === la.test(d)) return 115 === (b = d.substring(d.indexOf(':') + 1)).charCodeAt(0) ? P(d.replace('stretch', 'fill-available'), c, e, h).replace(':fill-available', ':stretch') : a.replace(b, '-webkit-' + b) + a.replace(b, '-moz-' + b.replace('fill-', '')) + a;
        break;
      case 962:
        if (a = '-webkit-' + a + (102 === a.charCodeAt(5) ? '-ms-' + a : '') + a, 211 === e + h && 105 === a.charCodeAt(13) && 0 < a.indexOf('transform', 10)) return a.substring(0, a.indexOf(';', 27) + 1).replace(ma, '$1-webkit-$2') + a;
    }
    return a;
  }
  function L(d, c) {
    var e = d.indexOf(1 === c ? ':' : '{'),
      h = d.substring(0, 3 !== c ? e : 10);
    e = d.substring(e + 1, d.length - 1);
    return R(2 !== c ? h : h.replace(na, '$1'), e, c);
  }
  function ea(d, c) {
    var e = P(c, c.charCodeAt(0), c.charCodeAt(1), c.charCodeAt(2));
    return e !== c + ';' ? e.replace(oa, ' or ($1)').substring(4) : '(' + c + ')';
  }
  function H(d, c, e, h, a, m, b, v, n, q) {
    for (var g = 0, x = c, w; g < A; ++g) {
      switch (w = S[g].call(B, d, x, e, h, a, m, b, v, n, q)) {
        case void 0:
        case !1:
        case !0:
        case null:
          break;
        default:
          x = w;
      }
    }
    if (x !== c) return x;
  }
  function T(d) {
    switch (d) {
      case void 0:
      case null:
        A = S.length = 0;
        break;
      default:
        if ('function' === typeof d) S[A++] = d;else if ('object' === typeof d) for (var c = 0, e = d.length; c < e; ++c) {
          T(d[c]);
        } else Y = !!d | 0;
    }
    return T;
  }
  function U(d) {
    d = d.prefix;
    void 0 !== d && (R = null, d ? 'function' !== typeof d ? w = 1 : (w = 2, R = d) : w = 0);
    return U;
  }
  function B(d, c) {
    var e = d;
    33 > e.charCodeAt(0) && (e = e.trim());
    V = e;
    e = [V];
    if (0 < A) {
      var h = H(-1, c, e, e, D, z, 0, 0, 0, 0);
      void 0 !== h && 'string' === typeof h && (c = h);
    }
    var a = M(O, e, c, 0, 0);
    0 < A && (h = H(-2, a, e, e, D, z, a.length, 0, 0, 0), void 0 !== h && (a = h));
    V = '';
    E = 0;
    z = D = 1;
    return a;
  }
  var ca = /^\0+/g,
    N = /[\0\r\f]/g,
    aa = /: */g,
    ka = /zoo|gra/,
    ma = /([,: ])(transform)/g,
    ia = /,\r+?/g,
    F = /([\t\r\n ])*\f?&/g,
    fa = /@(k\w+)\s*(\S*)\s*/,
    Q = /::(place)/g,
    ha = /:(read-only)/g,
    G = /[svh]\w+-[tblr]{2}/,
    da = /\(\s*(.*)\s*\)/g,
    oa = /([\s\S]*?);/g,
    ba = /-self|flex-/g,
    na = /[^]*?(:[rp][el]a[\w-]+)[^]*/,
    la = /stretch|:\s*\w+\-(?:conte|avail)/,
    ja = /([^-])(image-set\()/,
    z = 1,
    D = 1,
    E = 0,
    w = 1,
    O = [],
    S = [],
    A = 0,
    R = null,
    Y = 0,
    V = '';
  B.use = T;
  B.set = U;
  void 0 !== W && U(W);
  return B;
}

var unitlessKeys = {
  animationIterationCount: 1,
  borderImageOutset: 1,
  borderImageSlice: 1,
  borderImageWidth: 1,
  boxFlex: 1,
  boxFlexGroup: 1,
  boxOrdinalGroup: 1,
  columnCount: 1,
  columns: 1,
  flex: 1,
  flexGrow: 1,
  flexPositive: 1,
  flexShrink: 1,
  flexNegative: 1,
  flexOrder: 1,
  gridRow: 1,
  gridRowEnd: 1,
  gridRowSpan: 1,
  gridRowStart: 1,
  gridColumn: 1,
  gridColumnEnd: 1,
  gridColumnSpan: 1,
  gridColumnStart: 1,
  msGridRow: 1,
  msGridRowSpan: 1,
  msGridColumn: 1,
  msGridColumnSpan: 1,
  fontWeight: 1,
  lineHeight: 1,
  opacity: 1,
  order: 1,
  orphans: 1,
  tabSize: 1,
  widows: 1,
  zIndex: 1,
  zoom: 1,
  WebkitLineClamp: 1,
  // SVG-related properties
  fillOpacity: 1,
  floodOpacity: 1,
  stopOpacity: 1,
  strokeDasharray: 1,
  strokeDashoffset: 1,
  strokeMiterlimit: 1,
  strokeOpacity: 1,
  strokeWidth: 1
};

function memoize(fn) {
  var cache = Object.create(null);
  return function (arg) {
    if (cache[arg] === undefined) cache[arg] = fn(arg);
    return cache[arg];
  };
}

var reactPropsRegex = /^((children|dangerouslySetInnerHTML|key|ref|autoFocus|defaultValue|defaultChecked|innerHTML|suppressContentEditableWarning|suppressHydrationWarning|valueLink|abbr|accept|acceptCharset|accessKey|action|allow|allowUserMedia|allowPaymentRequest|allowFullScreen|allowTransparency|alt|async|autoComplete|autoPlay|capture|cellPadding|cellSpacing|challenge|charSet|checked|cite|classID|className|cols|colSpan|content|contentEditable|contextMenu|controls|controlsList|coords|crossOrigin|data|dateTime|decoding|default|defer|dir|disabled|disablePictureInPicture|download|draggable|encType|enterKeyHint|form|formAction|formEncType|formMethod|formNoValidate|formTarget|frameBorder|headers|height|hidden|high|href|hrefLang|htmlFor|httpEquiv|id|inputMode|integrity|is|keyParams|keyType|kind|label|lang|list|loading|loop|low|marginHeight|marginWidth|max|maxLength|media|mediaGroup|method|min|minLength|multiple|muted|name|nonce|noValidate|open|optimum|pattern|placeholder|playsInline|poster|preload|profile|radioGroup|readOnly|referrerPolicy|rel|required|reversed|role|rows|rowSpan|sandbox|scope|scoped|scrolling|seamless|selected|shape|size|sizes|slot|span|spellCheck|src|srcDoc|srcLang|srcSet|start|step|style|summary|tabIndex|target|title|translate|type|useMap|value|width|wmode|wrap|about|datatype|inlist|prefix|property|resource|typeof|vocab|autoCapitalize|autoCorrect|autoSave|color|incremental|fallback|inert|itemProp|itemScope|itemType|itemID|itemRef|on|option|results|security|unselectable|accentHeight|accumulate|additive|alignmentBaseline|allowReorder|alphabetic|amplitude|arabicForm|ascent|attributeName|attributeType|autoReverse|azimuth|baseFrequency|baselineShift|baseProfile|bbox|begin|bias|by|calcMode|capHeight|clip|clipPathUnits|clipPath|clipRule|colorInterpolation|colorInterpolationFilters|colorProfile|colorRendering|contentScriptType|contentStyleType|cursor|cx|cy|d|decelerate|descent|diffuseConstant|direction|display|divisor|dominantBaseline|dur|dx|dy|edgeMode|elevation|enableBackground|end|exponent|externalResourcesRequired|fill|fillOpacity|fillRule|filter|filterRes|filterUnits|floodColor|floodOpacity|focusable|fontFamily|fontSize|fontSizeAdjust|fontStretch|fontStyle|fontVariant|fontWeight|format|from|fr|fx|fy|g1|g2|glyphName|glyphOrientationHorizontal|glyphOrientationVertical|glyphRef|gradientTransform|gradientUnits|hanging|horizAdvX|horizOriginX|ideographic|imageRendering|in|in2|intercept|k|k1|k2|k3|k4|kernelMatrix|kernelUnitLength|kerning|keyPoints|keySplines|keyTimes|lengthAdjust|letterSpacing|lightingColor|limitingConeAngle|local|markerEnd|markerMid|markerStart|markerHeight|markerUnits|markerWidth|mask|maskContentUnits|maskUnits|mathematical|mode|numOctaves|offset|opacity|operator|order|orient|orientation|origin|overflow|overlinePosition|overlineThickness|panose1|paintOrder|pathLength|patternContentUnits|patternTransform|patternUnits|pointerEvents|points|pointsAtX|pointsAtY|pointsAtZ|preserveAlpha|preserveAspectRatio|primitiveUnits|r|radius|refX|refY|renderingIntent|repeatCount|repeatDur|requiredExtensions|requiredFeatures|restart|result|rotate|rx|ry|scale|seed|shapeRendering|slope|spacing|specularConstant|specularExponent|speed|spreadMethod|startOffset|stdDeviation|stemh|stemv|stitchTiles|stopColor|stopOpacity|strikethroughPosition|strikethroughThickness|string|stroke|strokeDasharray|strokeDashoffset|strokeLinecap|strokeLinejoin|strokeMiterlimit|strokeOpacity|strokeWidth|surfaceScale|systemLanguage|tableValues|targetX|targetY|textAnchor|textDecoration|textRendering|textLength|to|transform|u1|u2|underlinePosition|underlineThickness|unicode|unicodeBidi|unicodeRange|unitsPerEm|vAlphabetic|vHanging|vIdeographic|vMathematical|values|vectorEffect|version|vertAdvY|vertOriginX|vertOriginY|viewBox|viewTarget|visibility|widths|wordSpacing|writingMode|x|xHeight|x1|x2|xChannelSelector|xlinkActuate|xlinkArcrole|xlinkHref|xlinkRole|xlinkShow|xlinkTitle|xlinkType|xmlBase|xmlns|xmlnsXlink|xmlLang|xmlSpace|y|y1|y2|yChannelSelector|z|zoomAndPan|for|class|autofocus)|(([Dd][Aa][Tt][Aa]|[Aa][Rr][Ii][Aa]|x)-.*))$/; // https://esbench.com/bench/5bfee68a4cd7e6009ef61d23

var isPropValid = /* #__PURE__ */memoize(function (prop) {
  return reactPropsRegex.test(prop) || prop.charCodeAt(0) === 111
  /* o */ && prop.charCodeAt(1) === 110
  /* n */ && prop.charCodeAt(2) < 91;
}
/* Z+1 */);

var reactIs = reactIs$1.exports;

/**
 * Copyright 2015, Yahoo! Inc.
 * Copyrights licensed under the New BSD License. See the accompanying LICENSE file for terms.
 */
var REACT_STATICS = {
  childContextTypes: true,
  contextType: true,
  contextTypes: true,
  defaultProps: true,
  displayName: true,
  getDefaultProps: true,
  getDerivedStateFromError: true,
  getDerivedStateFromProps: true,
  mixins: true,
  propTypes: true,
  type: true
};
var KNOWN_STATICS = {
  name: true,
  length: true,
  prototype: true,
  caller: true,
  callee: true,
  arguments: true,
  arity: true
};
var FORWARD_REF_STATICS = {
  '$$typeof': true,
  render: true,
  defaultProps: true,
  displayName: true,
  propTypes: true
};
var MEMO_STATICS = {
  '$$typeof': true,
  compare: true,
  defaultProps: true,
  displayName: true,
  propTypes: true,
  type: true
};
var TYPE_STATICS = {};
TYPE_STATICS[reactIs.ForwardRef] = FORWARD_REF_STATICS;
TYPE_STATICS[reactIs.Memo] = MEMO_STATICS;
function getStatics(component) {
  // React v16.11 and below
  if (reactIs.isMemo(component)) {
    return MEMO_STATICS;
  } // React v16.12 and above

  return TYPE_STATICS[component['$$typeof']] || REACT_STATICS;
}
var defineProperty = Object.defineProperty;
var getOwnPropertyNames = Object.getOwnPropertyNames;
var getOwnPropertySymbols = Object.getOwnPropertySymbols;
var getOwnPropertyDescriptor = Object.getOwnPropertyDescriptor;
var getPrototypeOf = Object.getPrototypeOf;
var objectPrototype = Object.prototype;
function hoistNonReactStatics(targetComponent, sourceComponent, blacklist) {
  if (typeof sourceComponent !== 'string') {
    // don't hoist over string (html) components
    if (objectPrototype) {
      var inheritedComponent = getPrototypeOf(sourceComponent);
      if (inheritedComponent && inheritedComponent !== objectPrototype) {
        hoistNonReactStatics(targetComponent, inheritedComponent, blacklist);
      }
    }
    var keys = getOwnPropertyNames(sourceComponent);
    if (getOwnPropertySymbols) {
      keys = keys.concat(getOwnPropertySymbols(sourceComponent));
    }
    var targetStatics = getStatics(targetComponent);
    var sourceStatics = getStatics(sourceComponent);
    for (var i = 0; i < keys.length; ++i) {
      var key = keys[i];
      if (!KNOWN_STATICS[key] && !(blacklist && blacklist[key]) && !(sourceStatics && sourceStatics[key]) && !(targetStatics && targetStatics[key])) {
        var descriptor = getOwnPropertyDescriptor(sourceComponent, key);
        try {
          // Avoid failures from read-only properties
          defineProperty(targetComponent, key, descriptor);
        } catch (e) {}
      }
    }
  }
  return targetComponent;
}
var hoistNonReactStatics_cjs = hoistNonReactStatics;

function y() {
  return (y = Object.assign || function (e) {
    for (var t = 1; t < arguments.length; t++) {
      var n = arguments[t];
      for (var r in n) Object.prototype.hasOwnProperty.call(n, r) && (e[r] = n[r]);
    }
    return e;
  }).apply(this, arguments);
}
var v = function (e, t) {
    for (var n = [e[0]], r = 0, o = t.length; r < o; r += 1) n.push(t[r], e[r + 1]);
    return n;
  },
  g = function (t) {
    return null !== t && "object" == typeof t && "[object Object]" === (t.toString ? t.toString() : Object.prototype.toString.call(t)) && !reactIs$1.exports.typeOf(t);
  },
  S = Object.freeze([]),
  w = Object.freeze({});
function E(e) {
  return "function" == typeof e;
}
function b(e) {
  return "production" !== process.env.NODE_ENV && "string" == typeof e && e || e.displayName || e.name || "Component";
}
function _(e) {
  return e && "string" == typeof e.styledComponentId;
}
var N = "undefined" != typeof process && (process.env.REACT_APP_SC_ATTR || process.env.SC_ATTR) || "data-styled",
  C = "undefined" != typeof window && "HTMLElement" in window,
  I = Boolean("boolean" == typeof SC_DISABLE_SPEEDY ? SC_DISABLE_SPEEDY : "undefined" != typeof process && void 0 !== process.env.REACT_APP_SC_DISABLE_SPEEDY && "" !== process.env.REACT_APP_SC_DISABLE_SPEEDY ? "false" !== process.env.REACT_APP_SC_DISABLE_SPEEDY && process.env.REACT_APP_SC_DISABLE_SPEEDY : "undefined" != typeof process && void 0 !== process.env.SC_DISABLE_SPEEDY && "" !== process.env.SC_DISABLE_SPEEDY ? "false" !== process.env.SC_DISABLE_SPEEDY && process.env.SC_DISABLE_SPEEDY : "production" !== process.env.NODE_ENV),
  O = "production" !== process.env.NODE_ENV ? {
    1: "Cannot create styled-component for component: %s.\n\n",
    2: "Can't collect styles once you've consumed a `ServerStyleSheet`'s styles! `ServerStyleSheet` is a one off instance for each server-side render cycle.\n\n- Are you trying to reuse it across renders?\n- Are you accidentally calling collectStyles twice?\n\n",
    3: "Streaming SSR is only supported in a Node.js environment; Please do not try to call this method in the browser.\n\n",
    4: "The `StyleSheetManager` expects a valid target or sheet prop!\n\n- Does this error occur on the client and is your target falsy?\n- Does this error occur on the server and is the sheet falsy?\n\n",
    5: "The clone method cannot be used on the client!\n\n- Are you running in a client-like environment on the server?\n- Are you trying to run SSR on the client?\n\n",
    6: "Trying to insert a new style tag, but the given Node is unmounted!\n\n- Are you using a custom target that isn't mounted?\n- Does your document not have a valid head element?\n- Have you accidentally removed a style tag manually?\n\n",
    7: 'ThemeProvider: Please return an object from your "theme" prop function, e.g.\n\n```js\ntheme={() => ({})}\n```\n\n',
    8: 'ThemeProvider: Please make your "theme" prop an object.\n\n',
    9: "Missing document `<head>`\n\n",
    10: "Cannot find a StyleSheet instance. Usually this happens if there are multiple copies of styled-components loaded at once. Check out this issue for how to troubleshoot and fix the common cases where this situation can happen: https://github.com/styled-components/styled-components/issues/1941#issuecomment-417862021\n\n",
    11: "_This error was replaced with a dev-time warning, it will be deleted for v4 final._ [createGlobalStyle] received children which will not be rendered. Please use the component without passing children elements.\n\n",
    12: "It seems you are interpolating a keyframe declaration (%s) into an untagged string. This was supported in styled-components v3, but is not longer supported in v4 as keyframes are now injected on-demand. Please wrap your string in the css\\`\\` helper which ensures the styles are injected correctly. See https://www.styled-components.com/docs/api#css\n\n",
    13: "%s is not a styled component and cannot be referred to via component selector. See https://www.styled-components.com/docs/advanced#referring-to-other-components for more details.\n\n",
    14: 'ThemeProvider: "theme" prop is required.\n\n',
    15: "A stylis plugin has been supplied that is not named. We need a name for each plugin to be able to prevent styling collisions between different stylis configurations within the same app. Before you pass your plugin to `<StyleSheetManager stylisPlugins={[]}>`, please make sure each plugin is uniquely-named, e.g.\n\n```js\nObject.defineProperty(importedPlugin, 'name', { value: 'some-unique-name' });\n```\n\n",
    16: "Reached the limit of how many styled components may be created at group %s.\nYou may only create up to 1,073,741,824 components. If you're creating components dynamically,\nas for instance in your render method then you may be running into this limitation.\n\n",
    17: "CSSStyleSheet could not be found on HTMLStyleElement.\nHas styled-components' style tag been unmounted or altered by another script?\n"
  } : {};
function R() {
  for (var e = arguments.length <= 0 ? void 0 : arguments[0], t = [], n = 1, r = arguments.length; n < r; n += 1) t.push(n < 0 || arguments.length <= n ? void 0 : arguments[n]);
  return t.forEach(function (t) {
    e = e.replace(/%[a-z]/, t);
  }), e;
}
function D(e) {
  for (var t = arguments.length, n = new Array(t > 1 ? t - 1 : 0), r = 1; r < t; r++) n[r - 1] = arguments[r];
  throw "production" === process.env.NODE_ENV ? new Error("An error occurred. See https://git.io/JUIaE#" + e + " for more information." + (n.length > 0 ? " Args: " + n.join(", ") : "")) : new Error(R.apply(void 0, [O[e]].concat(n)).trim());
}
var j = function () {
    function e(e) {
      this.groupSizes = new Uint32Array(512), this.length = 512, this.tag = e;
    }
    var t = e.prototype;
    return t.indexOfGroup = function (e) {
      for (var t = 0, n = 0; n < e; n++) t += this.groupSizes[n];
      return t;
    }, t.insertRules = function (e, t) {
      if (e >= this.groupSizes.length) {
        for (var n = this.groupSizes, r = n.length, o = r; e >= o;) (o <<= 1) < 0 && D(16, "" + e);
        this.groupSizes = new Uint32Array(o), this.groupSizes.set(n), this.length = o;
        for (var s = r; s < o; s++) this.groupSizes[s] = 0;
      }
      for (var i = this.indexOfGroup(e + 1), a = 0, c = t.length; a < c; a++) this.tag.insertRule(i, t[a]) && (this.groupSizes[e]++, i++);
    }, t.clearGroup = function (e) {
      if (e < this.length) {
        var t = this.groupSizes[e],
          n = this.indexOfGroup(e),
          r = n + t;
        this.groupSizes[e] = 0;
        for (var o = n; o < r; o++) this.tag.deleteRule(n);
      }
    }, t.getGroup = function (e) {
      var t = "";
      if (e >= this.length || 0 === this.groupSizes[e]) return t;
      for (var n = this.groupSizes[e], r = this.indexOfGroup(e), o = r + n, s = r; s < o; s++) t += this.tag.getRule(s) + "/*!sc*/\n";
      return t;
    }, e;
  }(),
  T = new Map(),
  x = new Map(),
  k = 1,
  V = function (e) {
    if (T.has(e)) return T.get(e);
    for (; x.has(k);) k++;
    var t = k++;
    return "production" !== process.env.NODE_ENV && ((0 | t) < 0 || t > 1 << 30) && D(16, "" + t), T.set(e, t), x.set(t, e), t;
  },
  z = function (e) {
    return x.get(e);
  },
  B = function (e, t) {
    t >= k && (k = t + 1), T.set(e, t), x.set(t, e);
  },
  M = "style[" + N + '][data-styled-version="5.3.6"]',
  G = new RegExp("^" + N + '\\.g(\\d+)\\[id="([\\w\\d-]+)"\\].*?"([^"]*)'),
  L = function (e, t, n) {
    for (var r, o = n.split(","), s = 0, i = o.length; s < i; s++) (r = o[s]) && e.registerName(t, r);
  },
  F = function (e, t) {
    for (var n = (t.textContent || "").split("/*!sc*/\n"), r = [], o = 0, s = n.length; o < s; o++) {
      var i = n[o].trim();
      if (i) {
        var a = i.match(G);
        if (a) {
          var c = 0 | parseInt(a[1], 10),
            u = a[2];
          0 !== c && (B(u, c), L(e, u, a[3]), e.getTag().insertRules(c, r)), r.length = 0;
        } else r.push(i);
      }
    }
  },
  Y = function () {
    return "undefined" != typeof __webpack_nonce__ ? __webpack_nonce__ : null;
  },
  q = function (e) {
    var t = document.head,
      n = e || t,
      r = document.createElement("style"),
      o = function (e) {
        for (var t = e.childNodes, n = t.length; n >= 0; n--) {
          var r = t[n];
          if (r && 1 === r.nodeType && r.hasAttribute(N)) return r;
        }
      }(n),
      s = void 0 !== o ? o.nextSibling : null;
    r.setAttribute(N, "active"), r.setAttribute("data-styled-version", "5.3.6");
    var i = Y();
    return i && r.setAttribute("nonce", i), n.insertBefore(r, s), r;
  },
  H = function () {
    function e(e) {
      var t = this.element = q(e);
      t.appendChild(document.createTextNode("")), this.sheet = function (e) {
        if (e.sheet) return e.sheet;
        for (var t = document.styleSheets, n = 0, r = t.length; n < r; n++) {
          var o = t[n];
          if (o.ownerNode === e) return o;
        }
        D(17);
      }(t), this.length = 0;
    }
    var t = e.prototype;
    return t.insertRule = function (e, t) {
      try {
        return this.sheet.insertRule(t, e), this.length++, !0;
      } catch (e) {
        return !1;
      }
    }, t.deleteRule = function (e) {
      this.sheet.deleteRule(e), this.length--;
    }, t.getRule = function (e) {
      var t = this.sheet.cssRules[e];
      return void 0 !== t && "string" == typeof t.cssText ? t.cssText : "";
    }, e;
  }(),
  $ = function () {
    function e(e) {
      var t = this.element = q(e);
      this.nodes = t.childNodes, this.length = 0;
    }
    var t = e.prototype;
    return t.insertRule = function (e, t) {
      if (e <= this.length && e >= 0) {
        var n = document.createTextNode(t),
          r = this.nodes[e];
        return this.element.insertBefore(n, r || null), this.length++, !0;
      }
      return !1;
    }, t.deleteRule = function (e) {
      this.element.removeChild(this.nodes[e]), this.length--;
    }, t.getRule = function (e) {
      return e < this.length ? this.nodes[e].textContent : "";
    }, e;
  }(),
  W = function () {
    function e(e) {
      this.rules = [], this.length = 0;
    }
    var t = e.prototype;
    return t.insertRule = function (e, t) {
      return e <= this.length && (this.rules.splice(e, 0, t), this.length++, !0);
    }, t.deleteRule = function (e) {
      this.rules.splice(e, 1), this.length--;
    }, t.getRule = function (e) {
      return e < this.length ? this.rules[e] : "";
    }, e;
  }(),
  U = C,
  J = {
    isServer: !C,
    useCSSOMInjection: !I
  },
  X = function () {
    function e(e, t, n) {
      void 0 === e && (e = w), void 0 === t && (t = {}), this.options = y({}, J, {}, e), this.gs = t, this.names = new Map(n), this.server = !!e.isServer, !this.server && C && U && (U = !1, function (e) {
        for (var t = document.querySelectorAll(M), n = 0, r = t.length; n < r; n++) {
          var o = t[n];
          o && "active" !== o.getAttribute(N) && (F(e, o), o.parentNode && o.parentNode.removeChild(o));
        }
      }(this));
    }
    e.registerId = function (e) {
      return V(e);
    };
    var t = e.prototype;
    return t.reconstructWithOptions = function (t, n) {
      return void 0 === n && (n = !0), new e(y({}, this.options, {}, t), this.gs, n && this.names || void 0);
    }, t.allocateGSInstance = function (e) {
      return this.gs[e] = (this.gs[e] || 0) + 1;
    }, t.getTag = function () {
      return this.tag || (this.tag = (n = (t = this.options).isServer, r = t.useCSSOMInjection, o = t.target, e = n ? new W(o) : r ? new H(o) : new $(o), new j(e)));
      var e, t, n, r, o;
    }, t.hasNameForId = function (e, t) {
      return this.names.has(e) && this.names.get(e).has(t);
    }, t.registerName = function (e, t) {
      if (V(e), this.names.has(e)) this.names.get(e).add(t);else {
        var n = new Set();
        n.add(t), this.names.set(e, n);
      }
    }, t.insertRules = function (e, t, n) {
      this.registerName(e, t), this.getTag().insertRules(V(e), n);
    }, t.clearNames = function (e) {
      this.names.has(e) && this.names.get(e).clear();
    }, t.clearRules = function (e) {
      this.getTag().clearGroup(V(e)), this.clearNames(e);
    }, t.clearTag = function () {
      this.tag = void 0;
    }, t.toString = function () {
      return function (e) {
        for (var t = e.getTag(), n = t.length, r = "", o = 0; o < n; o++) {
          var s = z(o);
          if (void 0 !== s) {
            var i = e.names.get(s),
              a = t.getGroup(o);
            if (i && a && i.size) {
              var c = N + ".g" + o + '[id="' + s + '"]',
                u = "";
              void 0 !== i && i.forEach(function (e) {
                e.length > 0 && (u += e + ",");
              }), r += "" + a + c + '{content:"' + u + '"}/*!sc*/\n';
            }
          }
        }
        return r;
      }(this);
    }, e;
  }(),
  Z = /(a)(d)/gi,
  K = function (e) {
    return String.fromCharCode(e + (e > 25 ? 39 : 97));
  };
function Q(e) {
  var t,
    n = "";
  for (t = Math.abs(e); t > 52; t = t / 52 | 0) n = K(t % 52) + n;
  return (K(t % 52) + n).replace(Z, "$1-$2");
}
var ee = function (e, t) {
    for (var n = t.length; n;) e = 33 * e ^ t.charCodeAt(--n);
    return e;
  },
  te = function (e) {
    return ee(5381, e);
  };
function ne(e) {
  for (var t = 0; t < e.length; t += 1) {
    var n = e[t];
    if (E(n) && !_(n)) return !1;
  }
  return !0;
}
var re = te("5.3.6"),
  oe = function () {
    function e(e, t, n) {
      this.rules = e, this.staticRulesId = "", this.isStatic = "production" === process.env.NODE_ENV && (void 0 === n || n.isStatic) && ne(e), this.componentId = t, this.baseHash = ee(re, t), this.baseStyle = n, X.registerId(t);
    }
    return e.prototype.generateAndInjectStyles = function (e, t, n) {
      var r = this.componentId,
        o = [];
      if (this.baseStyle && o.push(this.baseStyle.generateAndInjectStyles(e, t, n)), this.isStatic && !n.hash) {
        if (this.staticRulesId && t.hasNameForId(r, this.staticRulesId)) o.push(this.staticRulesId);else {
          var s = _e(this.rules, e, t, n).join(""),
            i = Q(ee(this.baseHash, s) >>> 0);
          if (!t.hasNameForId(r, i)) {
            var a = n(s, "." + i, void 0, r);
            t.insertRules(r, i, a);
          }
          o.push(i), this.staticRulesId = i;
        }
      } else {
        for (var c = this.rules.length, u = ee(this.baseHash, n.hash), l = "", d = 0; d < c; d++) {
          var h = this.rules[d];
          if ("string" == typeof h) l += h, "production" !== process.env.NODE_ENV && (u = ee(u, h + d));else if (h) {
            var p = _e(h, e, t, n),
              f = Array.isArray(p) ? p.join("") : p;
            u = ee(u, f + d), l += f;
          }
        }
        if (l) {
          var m = Q(u >>> 0);
          if (!t.hasNameForId(r, m)) {
            var y = n(l, "." + m, void 0, r);
            t.insertRules(r, m, y);
          }
          o.push(m);
        }
      }
      return o.join(" ");
    }, e;
  }(),
  se = /^\s*\/\/.*$/gm,
  ie = [":", "[", ".", "#"];
function ae(e) {
  var t,
    n,
    r,
    o,
    s = void 0 === e ? w : e,
    i = s.options,
    a = void 0 === i ? w : i,
    c = s.plugins,
    u = void 0 === c ? S : c,
    l = new stylis_min(a),
    d = [],
    p = function (e) {
      function t(t) {
        if (t) try {
          e(t + "}");
        } catch (e) {}
      }
      return function (n, r, o, s, i, a, c, u, l, d) {
        switch (n) {
          case 1:
            if (0 === l && 64 === r.charCodeAt(0)) return e(r + ";"), "";
            break;
          case 2:
            if (0 === u) return r + "/*|*/";
            break;
          case 3:
            switch (u) {
              case 102:
              case 112:
                return e(o[0] + r), "";
              default:
                return r + (0 === d ? "/*|*/" : "");
            }
          case -2:
            r.split("/*|*/}").forEach(t);
        }
      };
    }(function (e) {
      d.push(e);
    }),
    f = function (e, r, s) {
      return 0 === r && -1 !== ie.indexOf(s[n.length]) || s.match(o) ? e : "." + t;
    };
  function m(e, s, i, a) {
    void 0 === a && (a = "&");
    var c = e.replace(se, ""),
      u = s && i ? i + " " + s + " { " + c + " }" : c;
    return t = a, n = s, r = new RegExp("\\" + n + "\\b", "g"), o = new RegExp("(\\" + n + "\\b){2,}"), l(i || !s ? "" : s, u);
  }
  return l.use([].concat(u, [function (e, t, o) {
    2 === e && o.length && o[0].lastIndexOf(n) > 0 && (o[0] = o[0].replace(r, f));
  }, p, function (e) {
    if (-2 === e) {
      var t = d;
      return d = [], t;
    }
  }])), m.hash = u.length ? u.reduce(function (e, t) {
    return t.name || D(15), ee(e, t.name);
  }, 5381).toString() : "", m;
}
var ce = React__default["default"].createContext();
  ce.Consumer;
  var le = React__default["default"].createContext(),
  de = (le.Consumer, new X()),
  he = ae();
function pe() {
  return React.useContext(ce) || de;
}
function fe() {
  return React.useContext(le) || he;
}
var ye = function () {
    function e(e, t) {
      var n = this;
      this.inject = function (e, t) {
        void 0 === t && (t = he);
        var r = n.name + t.hash;
        e.hasNameForId(n.id, r) || e.insertRules(n.id, r, t(n.rules, r, "@keyframes"));
      }, this.toString = function () {
        return D(12, String(n.name));
      }, this.name = e, this.id = "sc-keyframes-" + e, this.rules = t;
    }
    return e.prototype.getName = function (e) {
      return void 0 === e && (e = he), this.name + e.hash;
    }, e;
  }(),
  ve = /([A-Z])/,
  ge = /([A-Z])/g,
  Se = /^ms-/,
  we = function (e) {
    return "-" + e.toLowerCase();
  };
function Ee(e) {
  return ve.test(e) ? e.replace(ge, we).replace(Se, "-ms-") : e;
}
var be = function (e) {
  return null == e || !1 === e || "" === e;
};
function _e(e, n, r, o) {
  if (Array.isArray(e)) {
    for (var s, i = [], a = 0, c = e.length; a < c; a += 1) "" !== (s = _e(e[a], n, r, o)) && (Array.isArray(s) ? i.push.apply(i, s) : i.push(s));
    return i;
  }
  if (be(e)) return "";
  if (_(e)) return "." + e.styledComponentId;
  if (E(e)) {
    if ("function" != typeof (l = e) || l.prototype && l.prototype.isReactComponent || !n) return e;
    var u = e(n);
    return "production" !== process.env.NODE_ENV && reactIs$1.exports.isElement(u) && console.warn(b(e) + " is not a styled component and cannot be referred to via component selector. See https://www.styled-components.com/docs/advanced#referring-to-other-components for more details."), _e(u, n, r, o);
  }
  var l;
  return e instanceof ye ? r ? (e.inject(r, o), e.getName(o)) : e : g(e) ? function e(t, n) {
    var r,
      o,
      s = [];
    for (var i in t) t.hasOwnProperty(i) && !be(t[i]) && (Array.isArray(t[i]) && t[i].isCss || E(t[i]) ? s.push(Ee(i) + ":", t[i], ";") : g(t[i]) ? s.push.apply(s, e(t[i], i)) : s.push(Ee(i) + ": " + (r = i, null == (o = t[i]) || "boolean" == typeof o || "" === o ? "" : "number" != typeof o || 0 === o || r in unitlessKeys ? String(o).trim() : o + "px") + ";"));
    return n ? [n + " {"].concat(s, ["}"]) : s;
  }(e) : e.toString();
}
var Ne = function (e) {
  return Array.isArray(e) && (e.isCss = !0), e;
};
function Ae(e) {
  for (var t = arguments.length, n = new Array(t > 1 ? t - 1 : 0), r = 1; r < t; r++) n[r - 1] = arguments[r];
  return E(e) || g(e) ? Ne(_e(v(S, [e].concat(n)))) : 0 === n.length && 1 === e.length && "string" == typeof e[0] ? e : Ne(_e(v(e, n)));
}
var Ce = /invalid hook call/i,
  Ie = new Set(),
  Pe = function (e, t) {
    if ("production" !== process.env.NODE_ENV) {
      var n = "The component " + e + (t ? ' with the id of "' + t + '"' : "") + " has been created dynamically.\nYou may see this warning because you've called styled inside another component.\nTo resolve this only create new StyledComponents outside of any render method and function component.",
        r = console.error;
      try {
        var o = !0;
        console.error = function (e) {
          if (Ce.test(e)) o = !1, Ie.delete(n);else {
            for (var t = arguments.length, s = new Array(t > 1 ? t - 1 : 0), i = 1; i < t; i++) s[i - 1] = arguments[i];
            r.apply(void 0, [e].concat(s));
          }
        }, React.useRef(), o && !Ie.has(n) && (console.warn(n), Ie.add(n));
      } catch (e) {
        Ce.test(e.message) && Ie.delete(n);
      } finally {
        console.error = r;
      }
    }
  },
  Oe = function (e, t, n) {
    return void 0 === n && (n = w), e.theme !== n.theme && e.theme || t || n.theme;
  },
  Re = /[!"#$%&'()*+,./:;<=>?@[\\\]^`{|}~-]+/g,
  De = /(^-|-$)/g;
function je(e) {
  return e.replace(Re, "-").replace(De, "");
}
var Te = function (e) {
  return Q(te(e) >>> 0);
};
function xe(e) {
  return "string" == typeof e && ("production" === process.env.NODE_ENV || e.charAt(0) === e.charAt(0).toLowerCase());
}
var ke = function (e) {
    return "function" == typeof e || "object" == typeof e && null !== e && !Array.isArray(e);
  },
  Ve = function (e) {
    return "__proto__" !== e && "constructor" !== e && "prototype" !== e;
  };
function ze(e, t, n) {
  var r = e[n];
  ke(t) && ke(r) ? Be(r, t) : e[n] = t;
}
function Be(e) {
  for (var t = arguments.length, n = new Array(t > 1 ? t - 1 : 0), r = 1; r < t; r++) n[r - 1] = arguments[r];
  for (var o = 0, s = n; o < s.length; o++) {
    var i = s[o];
    if (ke(i)) for (var a in i) Ve(a) && ze(e, i[a], a);
  }
  return e;
}
var Me = React__default["default"].createContext();
  Me.Consumer;
var Fe = {};
function Ye(e, t, n) {
  var o = _(e),
    i = !xe(e),
    a = t.attrs,
    c = void 0 === a ? S : a,
    d = t.componentId,
    h = void 0 === d ? function (e, t) {
      var n = "string" != typeof e ? "sc" : je(e);
      Fe[n] = (Fe[n] || 0) + 1;
      var r = n + "-" + Te("5.3.6" + n + Fe[n]);
      return t ? t + "-" + r : r;
    }(t.displayName, t.parentComponentId) : d,
    p = t.displayName,
    v = void 0 === p ? function (e) {
      return xe(e) ? "styled." + e : "Styled(" + b(e) + ")";
    }(e) : p,
    g = t.displayName && t.componentId ? je(t.displayName) + "-" + t.componentId : t.componentId || h,
    N = o && e.attrs ? Array.prototype.concat(e.attrs, c).filter(Boolean) : c,
    A = t.shouldForwardProp;
  o && e.shouldForwardProp && (A = t.shouldForwardProp ? function (n, r, o) {
    return e.shouldForwardProp(n, r, o) && t.shouldForwardProp(n, r, o);
  } : e.shouldForwardProp);
  var C,
    I = new oe(n, g, o ? e.componentStyle : void 0),
    P = I.isStatic && 0 === c.length,
    O = function (e, t) {
      return function (e, t, n, r) {
        var o = e.attrs,
          i = e.componentStyle,
          a = e.defaultProps,
          c = e.foldedComponentIds,
          d = e.shouldForwardProp,
          h = e.styledComponentId,
          p = e.target;
        "production" !== process.env.NODE_ENV && React.useDebugValue(h);
        var m = function (e, t, n) {
            void 0 === e && (e = w);
            var r = y({}, t, {
                theme: e
              }),
              o = {};
            return n.forEach(function (e) {
              var t,
                n,
                s,
                i = e;
              for (t in E(i) && (i = i(r)), i) r[t] = o[t] = "className" === t ? (n = o[t], s = i[t], n && s ? n + " " + s : n || s) : i[t];
            }), [r, o];
          }(Oe(t, React.useContext(Me), a) || w, t, o),
          v = m[0],
          g = m[1],
          S = function (e, t, n, r) {
            var o = pe(),
              s = fe(),
              i = t ? e.generateAndInjectStyles(w, o, s) : e.generateAndInjectStyles(n, o, s);
            return "production" !== process.env.NODE_ENV && React.useDebugValue(i), "production" !== process.env.NODE_ENV && !t && r && r(i), i;
          }(i, r, v, "production" !== process.env.NODE_ENV ? e.warnTooManyClasses : void 0),
          b = n,
          _ = g.$as || t.$as || g.as || t.as || p,
          N = xe(_),
          A = g !== t ? y({}, t, {}, g) : t,
          C = {};
        for (var I in A) "$" !== I[0] && "as" !== I && ("forwardedAs" === I ? C.as = A[I] : (d ? d(I, isPropValid, _) : !N || isPropValid(I)) && (C[I] = A[I]));
        return t.style && g.style !== t.style && (C.style = y({}, t.style, {}, g.style)), C.className = Array.prototype.concat(c, h, S !== h ? S : null, t.className, g.className).filter(Boolean).join(" "), C.ref = b, React.createElement(_, C);
      }(C, e, t, P);
    };
  return O.displayName = v, (C = React__default["default"].forwardRef(O)).attrs = N, C.componentStyle = I, C.displayName = v, C.shouldForwardProp = A, C.foldedComponentIds = o ? Array.prototype.concat(e.foldedComponentIds, e.styledComponentId) : S, C.styledComponentId = g, C.target = o ? e.target : e, C.withComponent = function (e) {
    var r = t.componentId,
      o = function (e, t) {
        if (null == e) return {};
        var n,
          r,
          o = {},
          s = Object.keys(e);
        for (r = 0; r < s.length; r++) n = s[r], t.indexOf(n) >= 0 || (o[n] = e[n]);
        return o;
      }(t, ["componentId"]),
      s = r && r + "-" + (xe(e) ? e : je(b(e)));
    return Ye(e, y({}, o, {
      attrs: N,
      componentId: s
    }), n);
  }, Object.defineProperty(C, "defaultProps", {
    get: function () {
      return this._foldedDefaultProps;
    },
    set: function (t) {
      this._foldedDefaultProps = o ? Be({}, e.defaultProps, t) : t;
    }
  }), "production" !== process.env.NODE_ENV && (Pe(v, g), C.warnTooManyClasses = function (e, t) {
    var n = {},
      r = !1;
    return function (o) {
      if (!r && (n[o] = !0, Object.keys(n).length >= 200)) {
        var s = t ? ' with the id of "' + t + '"' : "";
        console.warn("Over 200 classes were generated for component " + e + s + ".\nConsider using the attrs method, together with a style object for frequently changed styles.\nExample:\n  const Component = styled.div.attrs(props => ({\n    style: {\n      background: props.background,\n    },\n  }))`width: 100%;`\n\n  <Component />"), r = !0, n = {};
      }
    };
  }(v, g)), C.toString = function () {
    return "." + C.styledComponentId;
  }, i && hoistNonReactStatics_cjs(C, e, {
    attrs: !0,
    componentStyle: !0,
    displayName: !0,
    foldedComponentIds: !0,
    shouldForwardProp: !0,
    styledComponentId: !0,
    target: !0,
    withComponent: !0
  }), C;
}
var qe = function (e) {
  return function e(t, r, o) {
    if (void 0 === o && (o = w), !reactIs$1.exports.isValidElementType(r)) return D(1, String(r));
    var s = function () {
      return t(r, o, Ae.apply(void 0, arguments));
    };
    return s.withConfig = function (n) {
      return e(t, r, y({}, o, {}, n));
    }, s.attrs = function (n) {
      return e(t, r, y({}, o, {
        attrs: Array.prototype.concat(o.attrs, n).filter(Boolean)
      }));
    }, s;
  }(Ye, e);
};
["a", "abbr", "address", "area", "article", "aside", "audio", "b", "base", "bdi", "bdo", "big", "blockquote", "body", "br", "button", "canvas", "caption", "cite", "code", "col", "colgroup", "data", "datalist", "dd", "del", "details", "dfn", "dialog", "div", "dl", "dt", "em", "embed", "fieldset", "figcaption", "figure", "footer", "form", "h1", "h2", "h3", "h4", "h5", "h6", "head", "header", "hgroup", "hr", "html", "i", "iframe", "img", "input", "ins", "kbd", "keygen", "label", "legend", "li", "link", "main", "map", "mark", "marquee", "menu", "menuitem", "meta", "meter", "nav", "noscript", "object", "ol", "optgroup", "option", "output", "p", "param", "picture", "pre", "progress", "q", "rp", "rt", "ruby", "s", "samp", "script", "section", "select", "small", "source", "span", "strong", "style", "sub", "summary", "sup", "table", "tbody", "td", "textarea", "tfoot", "th", "thead", "time", "title", "tr", "track", "u", "ul", "var", "video", "wbr", "circle", "clipPath", "defs", "ellipse", "foreignObject", "g", "image", "line", "linearGradient", "marker", "mask", "path", "pattern", "polygon", "polyline", "radialGradient", "rect", "stop", "svg", "text", "textPath", "tspan"].forEach(function (e) {
  qe[e] = qe(e);
});
"production" !== process.env.NODE_ENV && "undefined" != typeof navigator && "ReactNative" === navigator.product && console.warn("It looks like you've imported 'styled-components' on React Native.\nPerhaps you're looking to import 'styled-components/native'?\nRead more about this at https://www.styled-components.com/docs/basics#react-native"), "production" !== process.env.NODE_ENV && "test" !== process.env.NODE_ENV && "undefined" != typeof window && (window["__styled-components-init__"] = window["__styled-components-init__"] || 0, 1 === window["__styled-components-init__"] && console.warn("It looks like there are several instances of 'styled-components' initialized in this application. This may cause dynamic styles to not render properly, errors during the rehydration process, a missing theme prop, and makes your application bigger without good reason.\n\nSee https://s-c.sh/2BAXzed for more info."), window["__styled-components-init__"] += 1);
var styled = qe;

var Container = styled.span(templateObject_1$4 || (templateObject_1$4 = __makeTemplateObject(["\n  &::after {\n    content: '';\n    box-sizing: border-box;\n    width: 34px;\n    height: 34px;\n    position: absolute;\n    top: calc(50% - 20px);\n    left: calc(50% - 20px);\n    border: 4px solid var(--tg-theme-secondary-bg-color);\n    border-top-color: var(--tg-theme-button-color);\n    border-radius: 50%;\n    border-width: 2px;\n    animation: rotation 1s linear infinite;\n    @keyframes rotation {\n      0% {\n        transform: rotate(0deg);\n      }\n      100% {\n        transform: rotate(359deg);\n      }\n    }\n  }\n"], ["\n  &::after {\n    content: '';\n    box-sizing: border-box;\n    width: 34px;\n    height: 34px;\n    position: absolute;\n    top: calc(50% - 20px);\n    left: calc(50% - 20px);\n    border: 4px solid var(--tg-theme-secondary-bg-color);\n    border-top-color: var(--tg-theme-button-color);\n    border-radius: 50%;\n    border-width: 2px;\n    animation: rotation 1s linear infinite;\n    @keyframes rotation {\n      0% {\n        transform: rotate(0deg);\n      }\n      100% {\n        transform: rotate(359deg);\n      }\n    }\n  }\n"])));
var Spinner = function () {
    return (React__default["default"].createElement(Container, null));
};
var templateObject_1$4;

var Timer = /** @class */ (function () {
    function Timer(type) {
        this.duration = 0;
        this.elapsed = 0;
        this.isActive = false;
        this.lastFrameTime = Date.now();
        this.type = type;
        this.onTick = function () { };
        this.onCompleted = function () { };
        switch (this.type) {
            case "countDown":
                this.tickCountdown();
                break;
            case "timer":
                this.tick();
        }
    }
    Timer.prototype.getTimeLeft = function () {
        var difference = this.duration - this.elapsed;
        return Math.max(0, difference);
    };
    Timer.prototype.pause = function () {
        this.isActive = false;
        return this;
    };
    Timer.prototype.reset = function () {
        this.elapsed = 0;
    };
    Timer.prototype.setDuration = function (seconds) {
        this.lastFrameTime = Date.now();
        this.duration = seconds;
        return this;
    };
    Timer.prototype.start = function () {
        this.isActive = true;
        return this;
    };
    Timer.prototype.stop = function () {
        this.pause();
        this.reset();
    };
    Timer.prototype.tick = function () {
        var currentFrameTime = Date.now();
        var deltaTime = currentFrameTime - this.lastFrameTime;
        this.lastFrameTime = currentFrameTime;
        if (this.isActive) {
            this.elapsed -= deltaTime / 1000;
            this.onTick(this.getTimeLeft());
        }
        window.requestAnimationFrame(this.tick.bind(this));
    };
    Timer.prototype.tickCountdown = function () {
        var currentFrameTime = Date.now();
        var deltaTime = currentFrameTime - this.lastFrameTime;
        this.lastFrameTime = currentFrameTime;
        if (this.isActive) {
            this.elapsed += deltaTime / 1000;
            this.onTick(this.getTimeLeft());
            if (this.getTimeLeft() <= 0) {
                this.pause();
                this.onCompleted();
            }
        }
        window.requestAnimationFrame(this.tickCountdown.bind(this));
    };
    return Timer;
}());
var Wrapper$2 = styled.div(templateObject_1$3 || (templateObject_1$3 = __makeTemplateObject(["\n  width: 100%;\n  height: 100vh;\n  display: flex;\n  align-items: center;\n  justify-content: center;\n  position: relative;\n  background-color: black;\n  opacity: 0;\n  transition: 0.5s;\n  animation: show 0.5s 1;\n  animation-fill-mode: forwards;\n  @keyframes show {\n    0%{\n      opacity: 0;\n    }\n    100%{\n      opacity: 1;\n    }\n"], ["\n  width: 100%;\n  height: 100vh;\n  display: flex;\n  align-items: center;\n  justify-content: center;\n  position: relative;\n  background-color: black;\n  opacity: 0;\n  transition: 0.5s;\n  animation: show 0.5s 1;\n  animation-fill-mode: forwards;\n  @keyframes show {\n    0%{\n      opacity: 0;\n    }\n    100%{\n      opacity: 1;\n    }\n"])));
var AdTag$1 = styled.div(templateObject_2$3 || (templateObject_2$3 = __makeTemplateObject(["\n  padding: 2px 4px 1px 4px;\n  border-radius: 6px;\n  gap: 10px;\n  background: #FF9500;\n  font-size: 11px;\n  font-weight: 510;\n  line-height: 13px;\n  letter-spacing: 0.05999999865889549px;\n  text-align: left;\n  display: flex;\n  justify-content: center;\n  color: #FFFFFF;\n  font-family: \"SF Pro\",sans-serif;\n  @font-face {\n    font-family: 'SF Pro';\n    src: local('SF Pro'), url(\"../../fonts/SFPro/SF-Pro.ttf\") format('truetype');\n  }\n"], ["\n  padding: 2px 4px 1px 4px;\n  border-radius: 6px;\n  gap: 10px;\n  background: #FF9500;\n  font-size: 11px;\n  font-weight: 510;\n  line-height: 13px;\n  letter-spacing: 0.05999999865889549px;\n  text-align: left;\n  display: flex;\n  justify-content: center;\n  color: #FFFFFF;\n  font-family: \"SF Pro\",sans-serif;\n  @font-face {\n    font-family: 'SF Pro';\n    src: local('SF Pro'), url(\"../../fonts/SFPro/SF-Pro.ttf\") format('truetype');\n  }\n"])));
var TopInfoWrapper = styled.div(templateObject_3$1 || (templateObject_3$1 = __makeTemplateObject(["\n  width: 306px;\n  display: flex;\n  flex-direction: row;\n  gap: 6px;\n"], ["\n  width: 306px;\n  display: flex;\n  flex-direction: row;\n  gap: 6px;\n"])));
var TopButtonWrapper = styled.div(templateObject_4$1 || (templateObject_4$1 = __makeTemplateObject(["\n\n  display: flex;\n  flex-direction: row;\n  justify-content: flex-end;\n"], ["\n\n  display: flex;\n  flex-direction: row;\n  justify-content: flex-end;\n"])));
var TopWrapper = styled.div(templateObject_5$1 || (templateObject_5$1 = __makeTemplateObject(["\n  position: absolute;\n  top: 0;\n  width: 95%;\n  margin: 12px;\n  display: flex;\n  flex-direction: row;\n  align-items: center;\n  justify-content: space-between;\n  gap: 10px;\n"], ["\n  position: absolute;\n  top: 0;\n  width: 95%;\n  margin: 12px;\n  display: flex;\n  flex-direction: row;\n  align-items: center;\n  justify-content: space-between;\n  gap: 10px;\n"])));
var TopButton$1 = styled.button(templateObject_6$1 || (templateObject_6$1 = __makeTemplateObject(["\n  border-radius: 50%;\n  padding: 0;\n  min-width: 28px;\n  min-height: 28px;\n  width: 28px;\n  height: 28px;\n  margin-left: 6px;\n  display: flex;\n  align-items: center;\n  justify-content: center;\n  border: none;\n  outline: none;\n  background-color: var(--tg-theme-bg-color);\n  fill: var(--tg-theme-text-color);\n  cursor: ", ";\n"], ["\n  border-radius: 50%;\n  padding: 0;\n  min-width: 28px;\n  min-height: 28px;\n  width: 28px;\n  height: 28px;\n  margin-left: 6px;\n  display: flex;\n  align-items: center;\n  justify-content: center;\n  border: none;\n  outline: none;\n  background-color: var(--tg-theme-bg-color);\n  fill: var(--tg-theme-text-color);\n  cursor: ", ";\n"])), function (_a) {
    var disabled = _a.disabled;
    return (disabled ? "not-allowed" : "pointer");
});
var Text$1 = styled.p(templateObject_7$1 || (templateObject_7$1 = __makeTemplateObject(["\n  margin: 0;\n  color: var(--tg-theme-text-color);\n  white-space: nowrap;\n  overflow: hidden;\n  text-overflow: ellipsis;\n  font-size: 16px;\n  line-height: 17px;\n  font-weight: 510;\n  letter-spacing: -0.3100000023841858px;\n  text-align: left;\n  font-family: \"SF-Pro\",sans-serif;\n  @font-face {\n    font-family: 'SF-Pro';\n    src: local('SF-Pro'), url(\"../../fonts/SFPro/SF-Pro.ttf\") format('truetype');\n  }\n"], ["\n  margin: 0;\n  color: var(--tg-theme-text-color);\n  white-space: nowrap;\n  overflow: hidden;\n  text-overflow: ellipsis;\n  font-size: 16px;\n  line-height: 17px;\n  font-weight: 510;\n  letter-spacing: -0.3100000023841858px;\n  text-align: left;\n  font-family: \"SF-Pro\",sans-serif;\n  @font-face {\n    font-family: 'SF-Pro';\n    src: local('SF-Pro'), url(\"../../fonts/SFPro/SF-Pro.ttf\") format('truetype');\n  }\n"])));
styled.a(templateObject_8$1 || (templateObject_8$1 = __makeTemplateObject(["\n  text-decoration: none;\n"], ["\n  text-decoration: none;\n"])));
var VideoBlock = styled.video(templateObject_9$1 || (templateObject_9$1 = __makeTemplateObject(["\n  max-width: 100%;\n"], ["\n  max-width: 100%;\n"])));
var MuteIcom = styled(Memo$2)(templateObject_10$1 || (templateObject_10$1 = __makeTemplateObject(["\n  fill: var(--tg-theme-text-color);\n  min-width: 16px;\n"], ["\n  fill: var(--tg-theme-text-color);\n  min-width: 16px;\n"])));
var UnMuteIcom = styled(Memo$3)(templateObject_11$1 || (templateObject_11$1 = __makeTemplateObject(["\n  fill: var(--tg-theme-text-color);\n  min-width: 16px;\n"], ["\n  fill: var(--tg-theme-text-color);\n  min-width: 16px;\n"])));
styled(Memo)(templateObject_12$1 || (templateObject_12$1 = __makeTemplateObject(["\n  fill: white;\n  min-width: 16px;\n"], ["\n  fill: white;\n  min-width: 16px;\n"])));
var FullScreenAd = function (props) {
    props.link; var onClose = props.onClose, onClick = props.onClick, videoLink = props.videoLink, onMute = props.onMute, onError = props.onError, onPlayStart = props.onPlayStart, onPlayEnd = props.onPlayEnd;
    var _a = React.useState(true), isAdDisplayed = _a[0], setIsAdDisplayed = _a[1];
    var _b = React.useState(10), showAdTimer = _b[0], setShowAdTimer = _b[1];
    var _c = React.useState(true), isMuted = _c[0], setIsMuted = _c[1];
    var _d = React.useState(false), canPlay = _d[0], setCanPlay = _d[1];
    var adVideoRef = React.useRef(null);
    var onCanPlay = function () {
        if (adVideoRef.current != null) {
            adVideoRef.current.play();
            setCanPlay(true);
        }
    };
    React.useEffect(function () {
        var countDown = new Timer('countDown').setDuration(showAdTimer);
        if (!canPlay) {
            return;
        }
        countDown.onTick = function (time) {
            setShowAdTimer(Math.round(time));
        };
        countDown.start();
        return function () {
            countDown.stop();
        };
    }, [!canPlay]);
    var handleClose = function () {
        setIsAdDisplayed(false);
        onClose();
    };
    var handleEnded = function () {
        if (onPlayEnd) {
            onPlayEnd();
        }
        if (adVideoRef.current) {
            setIsAdDisplayed(false);
        }
    };
    var toggleMuteMode = function () {
        setIsMuted(function (prev) { return !prev; });
        onMute(isMuted);
    };
    var hanleAdClick = function () {
        onClick();
    };
    var onVideoError = function (e) {
        setIsAdDisplayed(false);
        onError(e);
    };
    if (!isAdDisplayed) {
        return null;
    }
    return (React__default["default"].createElement(Wrapper$2, null,
        React__default["default"].createElement(TopWrapper, null,
            React__default["default"].createElement(TopInfoWrapper, null,
                React__default["default"].createElement(AdTag$1, null, "Ad"),
                React__default["default"].createElement(Text$1, { hidden: showAdTimer < 1 || !canPlay },
                    "\u0421an be closed after ",
                    showAdTimer,
                    " sec")),
            React__default["default"].createElement(TopButtonWrapper, null,
                React__default["default"].createElement(TopButton$1, { onClick: toggleMuteMode }, isMuted ? React__default["default"].createElement(MuteIcom, null) : React__default["default"].createElement(UnMuteIcom, null)),
                React__default["default"].createElement(TopButton$1, { onClick: handleClose, disabled: showAdTimer > 0 },
                    React__default["default"].createElement(Memo, null)))),
        React__default["default"].createElement(VideoBlock, { ref: adVideoRef, src: videoLink, onError: function (e) { return onVideoError(e); }, onEnded: handleEnded, controls: false, muted: isMuted, autoPlay: true, onCanPlay: onCanPlay, onClick: hanleAdClick, onPlay: onPlayStart, playsInline: true },
            React__default["default"].createElement("source", { src: videoLink, type: "video/mp4" })),
        !canPlay ? React__default["default"].createElement("div", null,
            React__default["default"].createElement(Spinner, null)) : null));
};
var templateObject_1$3, templateObject_2$3, templateObject_3$1, templateObject_4$1, templateObject_5$1, templateObject_6$1, templateObject_7$1, templateObject_8$1, templateObject_9$1, templateObject_10$1, templateObject_11$1, templateObject_12$1;

var Placeholder = styled.div(templateObject_1$2 || (templateObject_1$2 = __makeTemplateObject(["\n    background-color: var(--tg-theme-bg-color);\n    opacity: 1;\n    display: block;\n    transition: opacity .2s var(--tg-theme-bg-color);\n    box-shadow: inset 0 0 0 2px var(--tg-theme-bg-color);\n    border-radius: 12px;\n"], ["\n    background-color: var(--tg-theme-bg-color);\n    opacity: 1;\n    display: block;\n    transition: opacity .2s var(--tg-theme-bg-color);\n    box-shadow: inset 0 0 0 2px var(--tg-theme-bg-color);\n    border-radius: 12px;\n"])));
var AdvertisementImage = styled.img(templateObject_2$2 || (templateObject_2$2 = __makeTemplateObject(["\n  border-radius: 12px;\n"], ["\n  border-radius: 12px;\n"])));
var ProgressiveImage = function (props) {
    var _a = React.useState(false), loaded = _a[0], setLoaded = _a[1];
    var src = props.src, width = props.width, height = props.height, alt = props.alt, className = props.className;
    React.useEffect(function () {
        if (!src) {
            return;
        }
        var image = new Image();
        image.onload = function () {
            setLoaded(true);
        };
        image.src = src;
    }, [src]);
    if (!src) {
        return (React__default["default"].createElement(Placeholder, null));
    }
    return (React__default["default"].createElement("div", { className: className },
        !loaded && React__default["default"].createElement(Placeholder, { style: { width: width, height: height } }),
        loaded && React__default["default"].createElement(AdvertisementImage, { src: src, width: width, height: height, alt: alt })));
};
var templateObject_1$2, templateObject_2$2;

var LinkWrapper = styled.div(templateObject_1$1 || (templateObject_1$1 = __makeTemplateObject(["\n  width: 89px;\n  height: 30px;\n  background-color: var(--tg-theme-button-color);\n  display: flex;\n  justify-content: center;\n  align-content: center;\n  border-radius: 24px;\n  cursor: pointer;\n  margin-top: auto;\n  margin-bottom: auto;\n  margin-right: 6px;\n"], ["\n  width: 89px;\n  height: 30px;\n  background-color: var(--tg-theme-button-color);\n  display: flex;\n  justify-content: center;\n  align-content: center;\n  border-radius: 24px;\n  cursor: pointer;\n  margin-top: auto;\n  margin-bottom: auto;\n  margin-right: 6px;\n"])));
var Wrapper$1 = styled.div(templateObject_2$1 || (templateObject_2$1 = __makeTemplateObject(["\n  padding: 10px;\n  width: 369px;\n  background-color: var(--tg-theme-bg-color);\n  border-radius: 24px;\n  box-shadow: 0px 0px 6px 0px rgba(0, 0, 0, 0.25);\n  display: flex;\n  text-decoration: none;\n"], ["\n  padding: 10px;\n  width: 369px;\n  background-color: var(--tg-theme-bg-color);\n  border-radius: 24px;\n  box-shadow: 0px 0px 6px 0px rgba(0, 0, 0, 0.25);\n  display: flex;\n  text-decoration: none;\n"])));
var HeaderWrapper = styled.div(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n  display: flex;\n  flex-direction: row;\n  align-items: center;\n"], ["\n  display: flex;\n  flex-direction: row;\n  align-items: center;\n"])));
var TopButton = styled.button(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n  width: 40px;\n  height: 30px;\n  padding: 6px;\n  border-radius: 15px;\n  display: flex;\n  justify-content: center;\n  align-content: center;\n  align-items: center;\n  margin-top: auto;\n  margin-bottom: auto;\n  border: none;\n  outline: none;\n  background-color: #7474801F;\n  color: #8E8E93;\n\n  cursor: ", ";\n"], ["\n  width: 40px;\n  height: 30px;\n  padding: 6px;\n  border-radius: 15px;\n  display: flex;\n  justify-content: center;\n  align-content: center;\n  align-items: center;\n  margin-top: auto;\n  margin-bottom: auto;\n  border: none;\n  outline: none;\n  background-color: #7474801F;\n  color: #8E8E93;\n\n  cursor: ", ";\n"])), function (_a) {
    var disabled = _a.disabled;
    return (disabled ? "not-allowed" : "pointer");
});
styled(ProgressiveImage)(templateObject_5 || (templateObject_5 = __makeTemplateObject(["\n  width: 64px;\n  height: 64px;\n  border-radius: 12px;\n"], ["\n  width: 64px;\n  height: 64px;\n  border-radius: 12px;\n"])));
var Htag = styled.h1(templateObject_6 || (templateObject_6 = __makeTemplateObject(["\n  color: var(--tg-theme-text-color);\n  font-weight: 510;\n  font-size: 17px;\n  line-height: 22px;\n  letter-spacing: -0.43px;\n  white-space: nowrap;\n  overflow: hidden;\n  text-overflow: ellipsis;\n  margin: 0;\n"], ["\n  color: var(--tg-theme-text-color);\n  font-weight: 510;\n  font-size: 17px;\n  line-height: 22px;\n  letter-spacing: -0.43px;\n  white-space: nowrap;\n  overflow: hidden;\n  text-overflow: ellipsis;\n  margin: 0;\n"])));
var InfoWrapper = styled.div(templateObject_7 || (templateObject_7 = __makeTemplateObject(["\n  min-width: 180px;\n  padding: 12px 8px 12px 12px;\n  display: flex;\n  flex-direction: column;\n  \n"], ["\n  min-width: 180px;\n  padding: 12px 8px 12px 12px;\n  display: flex;\n  flex-direction: column;\n  \n"])));
var AdTag = styled.div(templateObject_8 || (templateObject_8 = __makeTemplateObject(["\n  padding: 2px 4px 1px 4px;\n  border-radius: 6px;\n  margin-left: 4px;\n  gap: 10px;\n  background: #FF9500;\n  font-size: 11px;\n  font-weight: 510;\n  line-height: 13px;\n  letter-spacing: 0.05999999865889549px;\n  text-align: left;\n  display: flex;\n  justify-content: center;\n  color: #FFFFFF;\n"], ["\n  padding: 2px 4px 1px 4px;\n  border-radius: 6px;\n  margin-left: 4px;\n  gap: 10px;\n  background: #FF9500;\n  font-size: 11px;\n  font-weight: 510;\n  line-height: 13px;\n  letter-spacing: 0.05999999865889549px;\n  text-align: left;\n  display: flex;\n  justify-content: center;\n  color: #FFFFFF;\n"])));
var Text = styled.p(templateObject_9 || (templateObject_9 = __makeTemplateObject(["\n  margin: 0;\n  padding: 0;\n  color: #8E8E93;\n  font-size: 14px;\n  font-weight: 400;\n  line-height: 18px;\n  letter-spacing: -0.15px;\n  white-space: nowrap;\n  overflow: hidden;\n  text-overflow: ellipsis;\n"], ["\n  margin: 0;\n  padding: 0;\n  color: #8E8E93;\n  font-size: 14px;\n  font-weight: 400;\n  line-height: 18px;\n  letter-spacing: -0.15px;\n  white-space: nowrap;\n  overflow: hidden;\n  text-overflow: ellipsis;\n"])));
var CloseBtn = styled.div(templateObject_10 || (templateObject_10 = __makeTemplateObject(["\n  justify-content: center;\n  align-items: center;\n  margin-top: 2px;\n  @media only screen and (max-width: 1200px) and (min-width: 768px) {\n  display: list-item;\n  min-height: inherit;\n}\n"], ["\n  justify-content: center;\n  align-items: center;\n  margin-top: 2px;\n  @media only screen and (max-width: 1200px) and (min-width: 768px) {\n  display: list-item;\n  min-height: inherit;\n}\n"])));
var StyledTgAdsContainer = styled.div(templateObject_11 || (templateObject_11 = __makeTemplateObject(["\n  position: fixed;\n  bottom: 40px;\n  left: 50%;\n  transform: translate(-50%);\n  font-family: \"SF-Pro\",sans-serif;\n  @font-face {\n    font-family: 'SF-Pro';\n    src: local('SF-Pro'), url(\"../../fonts/SFPro/SF-Pro.ttf\") format('truetype');\n  }\n"], ["\n  position: fixed;\n  bottom: 40px;\n  left: 50%;\n  transform: translate(-50%);\n  font-family: \"SF-Pro\",sans-serif;\n  @font-face {\n    font-family: 'SF-Pro';\n    src: local('SF-Pro'), url(\"../../fonts/SFPro/SF-Pro.ttf\") format('truetype');\n  }\n"])));
var Link = styled.a(templateObject_12 || (templateObject_12 = __makeTemplateObject(["\n  font-size: 15px;\n  color: var(--tg-theme-button-text-color);\n  line-height: 18px;\n  font-weight: 590;\n  letter-spacing: -0.23px;\n  display: flex;\n  align-items: center;\n  gap: 4px;\n"], ["\n  font-size: 15px;\n  color: var(--tg-theme-button-text-color);\n  line-height: 18px;\n  font-weight: 590;\n  letter-spacing: -0.23px;\n  display: flex;\n  align-items: center;\n  gap: 4px;\n"])));
styled.h1(templateObject_13 || (templateObject_13 = __makeTemplateObject(["\n  font-size: 14px;\n  font-weight: 400;\n  letter-spacing: -0.4599999785423279px;\n  text-align: center;\n"], ["\n  font-size: 14px;\n  font-weight: 400;\n  letter-spacing: -0.4599999785423279px;\n  text-align: center;\n"])));
var BannerAd = function (props) {
    var title = props.title, text = props.text; props.link; var className = props.className, onClick = props.onClick, onClose = props.onClose, imageLink = props.imageLink, onError = props.onError;
    var _a = React.useState(true), isAdDisplayed = _a[0], setIsAdDisplayed = _a[1];
    var _b = React.useState(10), showAdTimer = _b[0], setShowAdTimer = _b[1];
    var handleClose = function () {
        setIsAdDisplayed(false);
        onClose();
    };
    React.useEffect(function () {
        var timerId;
        if (showAdTimer > 0) {
            timerId = setInterval(function () {
                setShowAdTimer(function (prev) { return prev - 1; });
            }, 1000);
        }
        return function () {
            clearInterval(timerId);
        };
    }, []);
    var hanleOpenBtnClick = function () {
        onClick();
    };
    if (!isAdDisplayed) {
        return null;
    }
    return (React__default["default"].createElement(StyledTgAdsContainer, null,
        React__default["default"].createElement(Wrapper$1, null,
            React__default["default"].createElement(ProgressiveImage, { src: imageLink, width: 64, height: 64, alt: "#", onError: onError }),
            React__default["default"].createElement(InfoWrapper, null,
                React__default["default"].createElement(HeaderWrapper, null,
                    React__default["default"].createElement(Htag, null, title),
                    React__default["default"].createElement(AdTag, null, "Ad")),
                React__default["default"].createElement(Text, null, text)),
            React__default["default"].createElement(LinkWrapper, { className: className, onClick: hanleOpenBtnClick },
                React__default["default"].createElement(Link, null, "Open")),
            React__default["default"].createElement(TopButton, { onClick: handleClose },
                "  ",
                React__default["default"].createElement(CloseBtn, null,
                    React__default["default"].createElement(Memo$1, null))))));
};
var templateObject_1$1, templateObject_2$1, templateObject_3, templateObject_4, templateObject_5, templateObject_6, templateObject_7, templateObject_8, templateObject_9, templateObject_10, templateObject_11, templateObject_12, templateObject_13;

var Wrapper = styled.div(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  display: flex;\n  align-items: center;\n  justify-content: center;\n  position: relative;\n"], ["\n  display: flex;\n  align-items: center;\n  justify-content: center;\n  position: relative;\n"])));
styled.div(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  width: 100%;\n  height: 99vh;\n  display: flex;\n  align-items: center;\n  justify-content: center;\n  position: relative;\n  @font-face {\n    font-family: 'SF-Pro';\n    src: local('SF-Pro'), url(\"./fonts/SFPro/SF-Pro.ttf\") format('truetype');\n  }\n"], ["\n  width: 100%;\n  height: 99vh;\n  display: flex;\n  align-items: center;\n  justify-content: center;\n  position: relative;\n  @font-face {\n    font-family: 'SF-Pro';\n    src: local('SF-Pro'), url(\"./fonts/SFPro/SF-Pro.ttf\") format('truetype');\n  }\n"])));
var TG_ADS_ADMIN_API_URL = "https://tgads-phi.vercel.app/api";
var TgAdsContainer = function (props) {
    var accessToken = props.accessToken, className = props.className, type = props.type, onClose = props.onClose, onClick = props.onClick, tgInitData = props.tgInitData, onError = props.onError, onLoadStart = props.onLoadStart, onLoadEnd = props.onLoadEnd, onPlayEnd = props.onPlayEnd, onPlayStart = props.onPlayStart;
    var _a = React.useState(false), isLoading = _a[0], setIsLoading = _a[1];
    var _b = React.useState(false), isCurrentPlatform = _b[0], setIsCurrentPlatform = _b[1];
    var _c = React.useState(null), currentAdData = _c[0], setCurrentAdData = _c[1];
    var _d = React.useState(0), timerTime = _d[0], setTimerTime = _d[1];
    var getIsCurrentPlatformState = React.useCallback(function () { return __awaiter(void 0, void 0, void 0, function () {
        var params, response, result, err_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 3, 4, 5]);
                    setIsLoading(true);
                    params = new URLSearchParams();
                    params.append("accessToken", accessToken);
                    params.append("adtype", type);
                    params.append("active", "true");
                    if (typeof window !== "undefined") {
                        params.append("clientRoute", window.location.origin);
                    }
                    return [4 /*yield*/, fetch("".concat(TG_ADS_ADMIN_API_URL, "/platforms/validate?") + params.toString())];
                case 1:
                    response = _a.sent();
                    if (!response.ok) {
                        throw new Error("HTTP Error: ".concat(response.status));
                    }
                    return [4 /*yield*/, response.json()];
                case 2:
                    result = _a.sent();
                    if (Boolean(result)) {
                        setIsCurrentPlatform(true);
                    }
                    setIsLoading(false);
                    return [3 /*break*/, 5];
                case 3:
                    err_1 = _a.sent();
                    console.log(err_1);
                    return [3 /*break*/, 5];
                case 4:
                    setIsLoading(false);
                    return [7 /*endfinally*/];
                case 5: return [2 /*return*/];
            }
        });
    }); }, []);
    var getAd = React.useCallback(function () { return __awaiter(void 0, void 0, void 0, function () {
        var params, response, result, e_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 3, 4, 5]);
                    setIsLoading(true);
                    params = new URLSearchParams();
                    params.append("accessToken", accessToken);
                    params.append("adtype", type.toString());
                    params.append("active", "true");
                    params.append("tgInitData", tgInitData);
                    if (onLoadStart) {
                        onLoadStart();
                    }
                    return [4 /*yield*/, fetch("".concat(TG_ADS_ADMIN_API_URL, "/notices/commercial?") + params.toString())];
                case 1:
                    response = _a.sent();
                    if (onLoadEnd) {
                        onLoadEnd();
                    }
                    if (!response.ok) {
                        throw new Error("HTTP Error: ".concat(response.status));
                    }
                    return [4 /*yield*/, response.json()];
                case 2:
                    result = _a.sent();
                    setCurrentAdData(result);
                    return [3 /*break*/, 5];
                case 3:
                    e_1 = _a.sent();
                    setIsLoading(false);
                    console.log(e_1);
                    return [3 /*break*/, 5];
                case 4:
                    setIsLoading(false);
                    return [7 /*endfinally*/];
                case 5: return [2 /*return*/];
            }
        });
    }); }, []);
    React.useEffect(function () {
        var timer = new Timer('timer');
        timer.onTick = function (time) {
            setTimerTime(time);
        };
        timer.start();
        return function () {
            timer.stop();
        };
    }, [currentAdData]);
    React.useEffect(function () {
        getIsCurrentPlatformState();
        if (isCurrentPlatform) {
            getAd();
        }
    }, [isCurrentPlatform]);
    var handleAdClick = React.useCallback(function () { return __awaiter(void 0, void 0, void 0, function () {
        var body, requestParams, response;
        var _a, _b;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    if (onClick) {
                        onClick();
                    }
                    if (!(typeof window !== "undefined")) return [3 /*break*/, 2];
                    body = {
                        platform: window.location.origin,
                        tgInitData: tgInitData,
                        ad: currentAdData.id,
                        action: "open",
                        time: timerTime
                    };
                    requestParams = {
                        method: "POST",
                        headers: {
                            "Content-Type": "application/json",
                        },
                        body: JSON.stringify(body),
                    };
                    return [4 /*yield*/, fetch("".concat(TG_ADS_ADMIN_API_URL, "/stats"), requestParams)];
                case 1:
                    response = _c.sent();
                    (_b = (_a = window === null || window === void 0 ? void 0 : window.Telegram) === null || _a === void 0 ? void 0 : _a.WebApp) === null || _b === void 0 ? void 0 : _b.openTelegramLink(currentAdData.link);
                    if (!response.ok) {
                        throw new Error("HTTP Error: ".concat(response.status));
                    }
                    if (!currentAdData) {
                        console.log("oops");
                        return [2 /*return*/];
                    }
                    _c.label = 2;
                case 2: return [2 /*return*/];
            }
        });
    }); }, [currentAdData]);
    var handleAdClose = React.useCallback(function () { return __awaiter(void 0, void 0, void 0, function () {
        var body, requestParams, response;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (onClose) {
                        onClose();
                    }
                    if (!currentAdData) {
                        console.log("oops");
                        return [2 /*return*/];
                    }
                    if (!(typeof window !== "undefined")) return [3 /*break*/, 2];
                    body = {
                        platform: window.location.origin,
                        tgInitData: tgInitData,
                        ad: currentAdData.id,
                        action: "close",
                        time: timerTime
                    };
                    requestParams = {
                        method: "POST",
                        headers: {
                            "Content-Type": "application/json",
                        },
                        body: JSON.stringify(body),
                    };
                    return [4 /*yield*/, fetch("".concat(TG_ADS_ADMIN_API_URL, "/stats"), requestParams)];
                case 1:
                    response = _a.sent();
                    if (!response.ok) {
                        throw new Error("HTTP Error: ".concat(response.status));
                    }
                    _a.label = 2;
                case 2:
                    setTimerTime(0);
                    return [2 /*return*/];
            }
        });
    }); }, [currentAdData]);
    var handleAdMute = React.useCallback(function (SoundStatus) { return __awaiter(void 0, void 0, void 0, function () {
        var action, body, requestParams, response;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!currentAdData) {
                        console.log("oops");
                        return [2 /*return*/];
                    }
                    action = "";
                    if (SoundStatus) {
                        action = "mute";
                    }
                    else {
                        action = "unmute";
                    }
                    if (!(typeof window !== "undefined")) return [3 /*break*/, 2];
                    body = {
                        platform: window.location.origin,
                        tgInitData: tgInitData,
                        ad: currentAdData.id,
                        action: action,
                        time: timerTime
                    };
                    requestParams = {
                        method: "POST",
                        headers: {
                            "Content-Type": "application/json",
                        },
                        body: JSON.stringify(body),
                    };
                    return [4 /*yield*/, fetch("".concat(TG_ADS_ADMIN_API_URL, "/stats"), requestParams)];
                case 1:
                    response = _a.sent();
                    if (!response.ok) {
                        throw new Error("HTTP Error: ".concat(response.status));
                    }
                    _a.label = 2;
                case 2: return [2 /*return*/];
            }
        });
    }); }, [currentAdData]);
    var handleError = React.useCallback(function (Error) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            if (!currentAdData) {
                console.log("oops");
                return [2 /*return*/];
            }
            if (onError) {
                return [2 /*return*/, onError(Error)];
            }
            return [2 /*return*/];
        });
    }); }, [currentAdData]);
    var currentAdContent = React.useMemo(function () {
        if (!currentAdData) {
            return null;
        }
        if (currentAdData.adtype === "banner") {
            if (typeof window !== "undefined") {
                var body = {
                    platform: window.location.origin,
                    tgInitData: tgInitData,
                    ad: currentAdData.id,
                    action: "view",
                };
                var requestParams = {
                    method: "POST",
                    headers: {
                        "Content-Type": "application/json",
                    },
                    body: JSON.stringify(body),
                };
                fetch("".concat(TG_ADS_ADMIN_API_URL, "/stats"), requestParams);
            }
            return (React__default["default"].createElement(Wrapper, null,
                React__default["default"].createElement(BannerAd, { title: currentAdData.title, text: currentAdData.text, link: currentAdData.link, className: className, onClick: handleAdClick, onClose: handleAdClose, imageLink: currentAdData.iconLink, onError: handleError })));
        }
        if (typeof window !== "undefined") {
            var body = {
                platform: window.location.origin,
                tgInitData: tgInitData,
                ad: currentAdData.id,
                action: "view",
            };
            var requestParams = {
                method: "POST",
                headers: {
                    "Content-Type": "application/json",
                },
                body: JSON.stringify(body),
            };
            fetch("".concat(TG_ADS_ADMIN_API_URL, "/stats"), requestParams);
        }
        return (React__default["default"].createElement(FullScreenAd, { onMute: handleAdMute, onClick: handleAdClick, link: currentAdData.link, onClose: handleAdClose, videoLink: currentAdData.videoLink, onError: handleError, onPlayStart: onPlayStart, onPlayEnd: onPlayEnd }));
    }, [currentAdData, type]);
    return !isLoading && currentAdData !== null ? currentAdContent : null;
};
var templateObject_1, templateObject_2;

exports.TgAdsContainer = TgAdsContainer;
